<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="A tool kit to manage many variant on desktop computer"><link href=https://SeqOIA-IT.github.io/variantplaner/usage/ rel=canonical><link rel=prev href=..><link href=../changelog/ rel=next><link rel=icon href=../assets/images/favicon.png><meta name=generator content="mkdocs-1.6.1, mkdocs-material-9.6.14"><title>Usage - variantplaner</title><link rel=stylesheet href=../assets/stylesheets/main.342714a4.min.css><link rel=stylesheet href=../assets/stylesheets/palette.06af60db.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=../assets/_markdown_exec_pyodide.css><link rel=stylesheet href=../css/timeago.css><link rel=stylesheet href=../assets/_mkdocstrings.css><link rel=stylesheet href=../css/material.css><link rel=stylesheet href=../css/mkdocstrings.css><script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script></head> <body dir=ltr data-md-color-scheme=slate data-md-color-primary=black data-md-color-accent=lime> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#variantplaner class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class="md-header md-header--shadow md-header--lifted" data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=.. title=variantplaner class="md-header__button md-logo" aria-label=variantplaner data-md-component=logo> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> variantplaner </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Usage </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme=slate data-md-color-primary=black data-md-color-accent=lime aria-label="Switch to light mode" type=radio name=__palette id=__palette_0> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3zm3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95zm-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31"/></svg> </label> <input class=md-option data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme=default data-md-color-primary=teal data-md-color-accent=purple aria-label="Switch to dark mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_0 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7 2.39 3.42C13.65 5.15 12.84 5 12 5s-1.65.15-2.39.42zM3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29zm.02 10 1.76-3.77a7.131 7.131 0 0 0 2.38 4.14zM20.65 7l-1.77 3.79a7.02 7.02 0 0 0-2.38-4.15zm-.01 10-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29zM12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44z"/></svg> </label> </form> <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg> </label> <nav class=md-search__options aria-label=Search> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap tabindex=0 data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/SeqOIA-IT/variantplaner title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg> </div> <div class=md-source__repository> SeqOIA-IT/variantplaner </div> </a> </div> </nav> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class=md-grid> <ul class=md-tabs__list> <li class="md-tabs__item md-tabs__item--active"> <a href=.. class=md-tabs__link> Home </a> </li> <li class=md-tabs__item> <a href=../reference/api/ class=md-tabs__link> API reference </a> </li> <li class=md-tabs__item> <a href=../contributing/ class=md-tabs__link> Development </a> </li> </ul> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=.. title=variantplaner class="md-nav__button md-logo" aria-label=variantplaner data-md-component=logo> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg> </a> variantplaner </label> <div class=md-nav__source> <a href=https://github.com/SeqOIA-IT/variantplaner title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg> </div> <div class=md-source__repository> SeqOIA-IT/variantplaner </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_1 checked> <label class=md-nav__link for=__nav_1 id=__nav_1_label tabindex> <span class=md-ellipsis> Home </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_1_label aria-expanded=true> <label class=md-nav__title for=__nav_1> <span class="md-nav__icon md-icon"></span> Home </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=.. class=md-nav__link> <span class=md-ellipsis> Overview </span> </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> <span class=md-ellipsis> Usage </span> <span class="md-nav__icon md-icon"></span> </label> <a href=./ class="md-nav__link md-nav__link--active"> <span class=md-ellipsis> Usage </span> </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#setup class=md-nav__link> <span class=md-ellipsis> Setup </span> </a> </li> <li class=md-nav__item> <a href=#download-data class=md-nav__link> <span class=md-ellipsis> Download data </span> </a> </li> <li class=md-nav__item> <a href=#variant-planner-presentation class=md-nav__link> <span class=md-ellipsis> Variant planner presentation </span> </a> </li> <li class=md-nav__item> <a href=#vcf2parquet class=md-nav__link> <span class=md-ellipsis> vcf2parquet </span> </a> </li> <li class=md-nav__item> <a href=#structuration-of-data class=md-nav__link> <span class=md-ellipsis> Structuration of data </span> </a> <nav class=md-nav aria-label="Structuration of data"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#merge-all-variant class=md-nav__link> <span class=md-ellipsis> Merge all variant </span> </a> </li> <li class=md-nav__item> <a href=#genotypes-structuration class=md-nav__link> <span class=md-ellipsis> Genotypes structuration </span> </a> </li> <li class=md-nav__item> <a href=#by-samples class=md-nav__link> <span class=md-ellipsis> By samples </span> </a> </li> <li class=md-nav__item> <a href=#by-variants class=md-nav__link> <span class=md-ellipsis> By variants </span> </a> </li> <li class=md-nav__item> <a href=#compute-transmission-mode class=md-nav__link> <span class=md-ellipsis> Compute transmission mode </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#add-annotations class=md-nav__link> <span class=md-ellipsis> Add annotations </span> </a> <nav class=md-nav aria-label="Add annotations"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#snpeff-annotations class=md-nav__link> <span class=md-ellipsis> Snpeff annotations </span> </a> </li> <li class=md-nav__item> <a href=#clinvar-annotations class=md-nav__link> <span class=md-ellipsis> Clinvar annotations </span> </a> </li> <li class=md-nav__item> <a href=#split-snpeff-annotations-in-parquet-column class=md-nav__link> <span class=md-ellipsis> Split snpeff annotations in parquet column </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#querying class=md-nav__link> <span class=md-ellipsis> Querying </span> </a> <nav class=md-nav aria-label=Querying> <ul class=md-nav__list> <li class=md-nav__item> <a href=#polars-cli class=md-nav__link> <span class=md-ellipsis> polars-cli </span> </a> <nav class=md-nav aria-label=polars-cli> <ul class=md-nav__list> <li class=md-nav__item> <a href=#count-variants class=md-nav__link> <span class=md-ellipsis> Count variants </span> </a> </li> <li class=md-nav__item> <a href=#filter-variants-from-annotations class=md-nav__link> <span class=md-ellipsis> Filter variants from annotations: </span> </a> </li> <li class=md-nav__item> <a href=#get-sample-have-variant class=md-nav__link> <span class=md-ellipsis> Get sample have variant </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#duckdb class=md-nav__link> <span class=md-ellipsis> duckdb </span> </a> <nav class=md-nav aria-label=duckdb> <ul class=md-nav__list> <li class=md-nav__item> <a href=#count-variants_1 class=md-nav__link> <span class=md-ellipsis> Count variants </span> </a> </li> <li class=md-nav__item> <a href=#filter-variants-from-annotations_1 class=md-nav__link> <span class=md-ellipsis> Filter variants from annotations: </span> </a> </li> <li class=md-nav__item> <a href=#get-sample-have-variant_1 class=md-nav__link> <span class=md-ellipsis> Get sample have variant </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#use-genotype-partition class=md-nav__link> <span class=md-ellipsis> Use genotype partition </span> </a> <nav class=md-nav aria-label="Use genotype partition"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#import class=md-nav__link> <span class=md-ellipsis> Import </span> </a> </li> <li class=md-nav__item> <a href=#get-variants class=md-nav__link> <span class=md-ellipsis> Get variants </span> </a> </li> <li class=md-nav__item> <a href=#add-annotations_1 class=md-nav__link> <span class=md-ellipsis> Add annotations </span> </a> </li> <li class=md-nav__item> <a href=#add-genotypes class=md-nav__link> <span class=md-ellipsis> Add genotypes </span> </a> </li> </ul> </nav> </li> </ul> </nav> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../changelog/ class=md-nav__link> <span class=md-ellipsis> Changelog </span> </a> </li> <li class=md-nav__item> <a href=../credits/ class=md-nav__link> <span class=md-ellipsis> Credits </span> </a> </li> <li class=md-nav__item> <a href=../license/ class=md-nav__link> <span class=md-ellipsis> License </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../reference/api/ class=md-nav__link> <span class=md-ellipsis> API reference </span> </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_3> <label class=md-nav__link for=__nav_3 id=__nav_3_label tabindex=0> <span class=md-ellipsis> Development </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_3_label aria-expanded=false> <label class=md-nav__title for=__nav_3> <span class="md-nav__icon md-icon"></span> Development </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../contributing/ class=md-nav__link> <span class=md-ellipsis> Contributing </span> </a> </li> <li class=md-nav__item> <a href=../code_of_conduct/ class=md-nav__link> <span class=md-ellipsis> Code of Conduct </span> </a> </li> <li class=md-nav__item> <a href=../coverage/ class=md-nav__link> <span class=md-ellipsis> Coverage report </span> </a> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#setup class=md-nav__link> <span class=md-ellipsis> Setup </span> </a> </li> <li class=md-nav__item> <a href=#download-data class=md-nav__link> <span class=md-ellipsis> Download data </span> </a> </li> <li class=md-nav__item> <a href=#variant-planner-presentation class=md-nav__link> <span class=md-ellipsis> Variant planner presentation </span> </a> </li> <li class=md-nav__item> <a href=#vcf2parquet class=md-nav__link> <span class=md-ellipsis> vcf2parquet </span> </a> </li> <li class=md-nav__item> <a href=#structuration-of-data class=md-nav__link> <span class=md-ellipsis> Structuration of data </span> </a> <nav class=md-nav aria-label="Structuration of data"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#merge-all-variant class=md-nav__link> <span class=md-ellipsis> Merge all variant </span> </a> </li> <li class=md-nav__item> <a href=#genotypes-structuration class=md-nav__link> <span class=md-ellipsis> Genotypes structuration </span> </a> </li> <li class=md-nav__item> <a href=#by-samples class=md-nav__link> <span class=md-ellipsis> By samples </span> </a> </li> <li class=md-nav__item> <a href=#by-variants class=md-nav__link> <span class=md-ellipsis> By variants </span> </a> </li> <li class=md-nav__item> <a href=#compute-transmission-mode class=md-nav__link> <span class=md-ellipsis> Compute transmission mode </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#add-annotations class=md-nav__link> <span class=md-ellipsis> Add annotations </span> </a> <nav class=md-nav aria-label="Add annotations"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#snpeff-annotations class=md-nav__link> <span class=md-ellipsis> Snpeff annotations </span> </a> </li> <li class=md-nav__item> <a href=#clinvar-annotations class=md-nav__link> <span class=md-ellipsis> Clinvar annotations </span> </a> </li> <li class=md-nav__item> <a href=#split-snpeff-annotations-in-parquet-column class=md-nav__link> <span class=md-ellipsis> Split snpeff annotations in parquet column </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#querying class=md-nav__link> <span class=md-ellipsis> Querying </span> </a> <nav class=md-nav aria-label=Querying> <ul class=md-nav__list> <li class=md-nav__item> <a href=#polars-cli class=md-nav__link> <span class=md-ellipsis> polars-cli </span> </a> <nav class=md-nav aria-label=polars-cli> <ul class=md-nav__list> <li class=md-nav__item> <a href=#count-variants class=md-nav__link> <span class=md-ellipsis> Count variants </span> </a> </li> <li class=md-nav__item> <a href=#filter-variants-from-annotations class=md-nav__link> <span class=md-ellipsis> Filter variants from annotations: </span> </a> </li> <li class=md-nav__item> <a href=#get-sample-have-variant class=md-nav__link> <span class=md-ellipsis> Get sample have variant </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#duckdb class=md-nav__link> <span class=md-ellipsis> duckdb </span> </a> <nav class=md-nav aria-label=duckdb> <ul class=md-nav__list> <li class=md-nav__item> <a href=#count-variants_1 class=md-nav__link> <span class=md-ellipsis> Count variants </span> </a> </li> <li class=md-nav__item> <a href=#filter-variants-from-annotations_1 class=md-nav__link> <span class=md-ellipsis> Filter variants from annotations: </span> </a> </li> <li class=md-nav__item> <a href=#get-sample-have-variant_1 class=md-nav__link> <span class=md-ellipsis> Get sample have variant </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#use-genotype-partition class=md-nav__link> <span class=md-ellipsis> Use genotype partition </span> </a> <nav class=md-nav aria-label="Use genotype partition"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#import class=md-nav__link> <span class=md-ellipsis> Import </span> </a> </li> <li class=md-nav__item> <a href=#get-variants class=md-nav__link> <span class=md-ellipsis> Get variants </span> </a> </li> <li class=md-nav__item> <a href=#add-annotations_1 class=md-nav__link> <span class=md-ellipsis> Add annotations </span> </a> </li> <li class=md-nav__item> <a href=#add-genotypes class=md-nav__link> <span class=md-ellipsis> Add genotypes </span> </a> </li> </ul> </nav> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <a href=https://github.com/SeqOIA-IT/variantplaner/edit/main/docs/usage.md title="Edit this page" class="md-content__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M10 20H6V4h7v5h5v3.1l2-2V8l-6-6H6c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h4zm10.2-7c.1 0 .3.1.4.2l1.3 1.3c.2.2.2.6 0 .8l-1 1-2.1-2.1 1-1c.1-.1.2-.2.4-.2m0 3.9L14.1 23H12v-2.1l6.1-6.1z"/></svg> </a> <a href=https://github.com/SeqOIA-IT/variantplaner/raw/main/docs/usage.md title="View source of this page" class="md-content__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 18c.56 0 1 .44 1 1s-.44 1-1 1-1-.44-1-1 .44-1 1-1m0-3c-2.73 0-5.06 1.66-6 4 .94 2.34 3.27 4 6 4s5.06-1.66 6-4c-.94-2.34-3.27-4-6-4m0 6.5a2.5 2.5 0 0 1-2.5-2.5 2.5 2.5 0 0 1 2.5-2.5 2.5 2.5 0 0 1 2.5 2.5 2.5 2.5 0 0 1-2.5 2.5M9.27 20H6V4h7v5h5v4.07c.7.08 1.36.25 2 .49V8l-6-6H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h4.5a8.2 8.2 0 0 1-1.23-2"/></svg> </a> <h1 id=variantplaner>VariantPlaner<a class=headerlink href=#variantplaner title="Permanent link">&para;</a></h1> <p>variantplaner is a set of tools that converts a large set of vcf's into an interoperable data structure efficiently.</p> <p>To show the capabilities of the <code>variantplaner</code>, we will use a small example.</p> <p>The purpose of this short tutorial is to present:</p> <ul> <li>how to convert vcf into a more suitable format</li> <li>how data can be restructured for querying</li> <li>how to integrate variant annotation databases</li> <li>how these different files can be used to obtain interesting biological information</li> </ul> <p>This tutorial suggests an organization of files, but you're under no obligation to follow it <code>variantplaner</code> is quite flexible in its organization.</p> <h2 id=setup>Setup<a class=headerlink href=#setup title="Permanent link">&para;</a></h2> <p>This tutorial assume you are on unix like system, you have python setup and you <a href=https://github.com/SeqOIA-IT/variantplaner#installation>install variantplaner</a></p> <p>Requirements list:</p> <ul> <li>curl</li> <li><a href=https://github.com/manojkarthick/pqrs>pqrs</a> (only for transmission computation, otherwise optional)</li> </ul> <p>Optional:</p> <ul> <li>gnu-parallel</li> </ul> <p>Quering dataset:</p> <ul> <li><a href=https://crates.io/crates/polars-cli>polars-cli</a></li> <li><a href=https://duckdb.org/ >duckdb</a></li> </ul> <h2 id=download-data>Download data<a class=headerlink href=#download-data title="Permanent link">&para;</a></h2> <div class=highlight><pre><span></span><code>mkdir<span class=w> </span>-p<span class=w> </span>vp_tuto/vcf/
<span class=nb>cd</span><span class=w> </span>vp_tuto
<span class=nv>URI_ROOT</span><span class=o>=</span><span class=s2>&quot;https://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/release&quot;</span>
curl<span class=w> </span><span class=si>${</span><span class=nv>URI_ROOT</span><span class=si>}</span>/NA12878_HG001/latest/GRCh38/HG001_GRCh38_1_22_v4.2.1_benchmark.vcf.gz<span class=w> </span>&gt;<span class=w> </span>vcf/HG001.vcf.gz
curl<span class=w> </span><span class=si>${</span><span class=nv>URI_ROOT</span><span class=si>}</span>/AshkenazimTrio/HG002_NA24385_son/latest/GRCh38/HG002_GRCh38_1_22_v4.2.1_benchmark.vcf.gz<span class=w> </span>&gt;<span class=w> </span>vcf/HG002.vcf.gz
curl<span class=w> </span><span class=si>${</span><span class=nv>URI_ROOT</span><span class=si>}</span>/AshkenazimTrio/HG003_NA24149_father/latest/GRCh38/HG003_GRCh38_1_22_v4.2.1_benchmark.vcf.gz<span class=w> </span>&gt;<span class=w> </span>vcf/HG003.vcf.gz
curl<span class=w> </span><span class=si>${</span><span class=nv>URI_ROOT</span><span class=si>}</span>/AshkenazimTrio/HG004_NA24143_mother/latest/GRCh38/HG004_GRCh38_1_22_v4.2.1_benchmark.vcf.gz<span class=w> </span>&gt;<span class=w> </span>vcf/HG004.vcf.gz
curl<span class=w> </span><span class=si>${</span><span class=nv>URI_ROOT</span><span class=si>}</span>/ChineseTrio/HG006_NA24694_father/latest/GRCh38/HG006_GRCh38_1_22_v4.2.1_benchmark.vcf.gz<span class=w> </span>&gt;<span class=w> </span>vcf/HG006.vcf.gz
curl<span class=w> </span><span class=si>${</span><span class=nv>URI_ROOT</span><span class=si>}</span>/ChineseTrio/HG007_NA24695_mother/latest/GRCh38/HG007_GRCh38_1_22_v4.2.1_benchmark.vcf.gz<span class=w> </span>&gt;<span class=w> </span>vcf/HG007.vcf.gz
</code></pre></div> <h2 id=variant-planner-presentation>Variant planner presentation<a class=headerlink href=#variant-planner-presentation title="Permanent link">&para;</a></h2> <p>variantplaner is a python module with command line tools, it is composed of several subcommands (they will be detailed later) but has two global options, one for parallelization and another for the level of verbosity you want.</p> <div class=highlight><pre><span></span><code>Usage: variantplaner [OPTIONS] COMMAND [ARGS]...

  Run VariantPlanner.

Options:
  -t, --threads INTEGER RANGE  Number of threads usable  [default: 1; x&gt;=0]
  -v, --verbose                Verbosity level  [0&lt;=x&lt;=4]
  --debug-info                 Get debug information
  -h, --help                   Show this message and exit.

Commands:
  metadata      Convert metadata file in parquet file.
  parquet2vcf   Convert variant parquet in vcf.
  struct        Subcommand to made struct operation on parquet file.
  transmission  Generate transmission of a genotype set.
  vcf2parquet   Convert a vcf in parquet.
</code></pre></div> <h2 id=vcf2parquet>vcf2parquet<a class=headerlink href=#vcf2parquet title="Permanent link">&para;</a></h2> <p>First step is to convert vcf data in <a href=https://en.wikipedia.org/wiki/Apache_Parquet>parquet</a>, it's a column oriented format with better performance than indexed vcf.</p> <p>We split vcf in two part on for variant information and another for genotype information.</p> <div class=highlight><pre><span></span><code>mkdir<span class=w> </span>-p<span class=w> </span>variants<span class=w> </span>genotypes/samples/
</code></pre></div> <div class=highlight><pre><span></span><code>for vcf_path in $(ls vcf/*.vcf.gz)
do
    sample_name=$(basename ${vcf_path} .vcf.gz)
    variantplaner -t 4 vcf2parquet -i ${vcf_path} \
    variants -o variants/${sample_name}.parquet \
    genotypes -o genotypes/samples/${sample_name}.parquet \
done
</code></pre></div> <p>We iterate over all vcf, variants are store in <code>variants/{sample_name}.parquet</code>, genotype information are store in <code>variants/{sample_name}.parquet</code>. Only genotypes with a format value equal to the <code>-f</code> parameter value are retained.</p> <details> <summary>gnu-parallel method</summary> <div class=highlight><pre><span></span><code>find<span class=w> </span>vcf<span class=w> </span>-type<span class=w> </span>f<span class=w> </span>-name<span class=w> </span>*.vcf<span class=w> </span>-exec<span class=w> </span>basename<span class=w> </span><span class=o>{}</span><span class=w> </span>.vcf<span class=w> </span><span class=se>\;</span><span class=w> </span><span class=p>|</span><span class=w> </span><span class=se>\</span>
parallel<span class=w> </span>variantplaner<span class=w> </span>-t<span class=w> </span><span class=m>2</span><span class=w> </span>vcf2parquet<span class=w> </span>-i<span class=w> </span>vcf/<span class=o>{}</span>.vcf<span class=w> </span><span class=se>\</span>
variants<span class=w> </span>-o<span class=w> </span>variants/<span class=o>{}</span>.parquet<span class=w> </span>genotypes<span class=w> </span>-o<span class=w> </span>genotypes/samples/<span class=o>{}</span>.parquet<span class=w> </span>-f<span class=w> </span>GT:PS:DP:ADALL:AD:GQ
</code></pre></div> </details> <p>Parquet variants file contains 5 column:</p> <ul> <li>pos: Position of variant</li> <li>ref: Reference sequence</li> <li>alt: Alternative sequence</li> <li>id: An hash of other value collision isn't check but highly improbable <a href=reference/variantplaner/normalization.md#variantplaner.normalization.add_variant_i>check api documentation</a></li> </ul> <div class="admonition danger"> <p class=admonition-title>Danger</p> <p>Column <code>id</code> value dependent on chromosome length and chromosome order in the VCF header.</p> <p>If you cannot be sure that the parameters are the same between each VCF file, you must create a chromosome length file. Section <a href=#add-annotations>add annotations</a> shows an example of this file.</p> </div> <details> <summary>variants parquet file content</summary> <p>You can inspect content of parquet file generate with pqrs <div class=highlight><pre><span></span><code>pqrs<span class=w> </span>head<span class=w> </span>variants/HG001.parquet
<span class=o>{</span>id:<span class=w> </span><span class=m>17886044532216650390</span>,<span class=w> </span>chr:<span class=w> </span><span class=m>1</span>,<span class=w> </span>pos:<span class=w> </span><span class=m>783006</span>,<span class=w> </span>ref:<span class=w> </span><span class=s2>&quot;A&quot;</span>,<span class=w> </span>alt:<span class=w> </span><span class=s2>&quot;G&quot;</span><span class=o>}</span>
<span class=o>{</span>id:<span class=w> </span><span class=m>7513336577790240873</span>,<span class=w> </span>chr:<span class=w> </span><span class=m>1</span>,<span class=w> </span>pos:<span class=w> </span><span class=m>783175</span>,<span class=w> </span>ref:<span class=w> </span><span class=s2>&quot;T&quot;</span>,<span class=w> </span>alt:<span class=w> </span><span class=s2>&quot;C&quot;</span><span class=o>}</span>
<span class=o>{</span>id:<span class=w> </span><span class=m>17987040642944149052</span>,<span class=w> </span>chr:<span class=w> </span><span class=m>1</span>,<span class=w> </span>pos:<span class=w> </span><span class=m>784860</span>,<span class=w> </span>ref:<span class=w> </span><span class=s2>&quot;T&quot;</span>,<span class=w> </span>alt:<span class=w> </span><span class=s2>&quot;C&quot;</span><span class=o>}</span>
<span class=o>{</span>id:<span class=w> </span><span class=m>10342734968077036194</span>,<span class=w> </span>chr:<span class=w> </span><span class=m>1</span>,<span class=w> </span>pos:<span class=w> </span><span class=m>785417</span>,<span class=w> </span>ref:<span class=w> </span><span class=s2>&quot;G&quot;</span>,<span class=w> </span>alt:<span class=w> </span><span class=s2>&quot;A&quot;</span><span class=o>}</span>
<span class=o>{</span>id:<span class=w> </span><span class=m>890514037559296207</span>,<span class=w> </span>chr:<span class=w> </span><span class=m>1</span>,<span class=w> </span>pos:<span class=w> </span><span class=m>797392</span>,<span class=w> </span>ref:<span class=w> </span><span class=s2>&quot;G&quot;</span>,<span class=w> </span>alt:<span class=w> </span><span class=s2>&quot;A&quot;</span><span class=o>}</span>
</code></pre></div></p> </details> <p>Parquet genotypes file contains column:</p> <ul> <li>id: Same as variant id</li> <li>gt: vcf GT value 1 -&gt; heterozygote 2 -&gt; homozygote (phasing information is lost)</li> <li>ps: Phase set in which this variant falls</li> <li>dp: vcf DP coverage of the variant for this sample</li> <li>adall: Net allele depths across all datasets</li> <li>ad: vcf AD per allele reads depth</li> <li>gq: vcf GQ quality of variant for this sample</li> </ul> <details> <summary>genotypes parquet file content</summary> <p>You can inspect content of parquet file generate with pqrs <div class=highlight><pre><span></span><code>pqrs<span class=w> </span>head<span class=w> </span>genotypes/samples/HG001.parquet
<span class=o>{</span>id:<span class=w> </span><span class=m>17886044532216650390</span>,<span class=w> </span>sample:<span class=w> </span><span class=s2>&quot;HG001&quot;</span>,<span class=w> </span>gt:<span class=w> </span><span class=m>2</span>,<span class=w> </span>ps:<span class=w> </span>null,<span class=w> </span>dp:<span class=w> </span><span class=m>652</span>,<span class=w> </span>adall:<span class=w> </span><span class=o>[</span><span class=m>16</span>,<span class=w> </span><span class=m>234</span><span class=o>]</span>,<span class=w> </span>ad:<span class=w> </span><span class=o>[</span><span class=m>0</span>,<span class=w> </span><span class=m>82</span><span class=o>]</span>,<span class=w> </span>gq:<span class=w> </span><span class=m>312</span><span class=o>}</span>
<span class=o>{</span>id:<span class=w> </span><span class=m>7513336577790240873</span>,<span class=w> </span>sample:<span class=w> </span><span class=s2>&quot;HG001&quot;</span>,<span class=w> </span>gt:<span class=w> </span><span class=m>2</span>,<span class=w> </span>ps:<span class=w> </span>null,<span class=w> </span>dp:<span class=w> </span><span class=m>639</span>,<span class=w> </span>adall:<span class=w> </span><span class=o>[</span><span class=m>0</span>,<span class=w> </span><span class=m>218</span><span class=o>]</span>,<span class=w> </span>ad:<span class=w> </span><span class=o>[</span><span class=m>0</span>,<span class=w> </span><span class=m>84</span><span class=o>]</span>,<span class=w> </span>gq:<span class=w> </span><span class=m>194</span><span class=o>}</span>
<span class=o>{</span>id:<span class=w> </span><span class=m>17987040642944149052</span>,<span class=w> </span>sample:<span class=w> </span><span class=s2>&quot;HG001&quot;</span>,<span class=w> </span>gt:<span class=w> </span><span class=m>2</span>,<span class=w> </span>ps:<span class=w> </span>null,<span class=w> </span>dp:<span class=w> </span><span class=m>901</span>,<span class=w> </span>adall:<span class=w> </span><span class=o>[</span><span class=m>105</span>,<span class=w> </span><span class=m>406</span><span class=o>]</span>,<span class=w> </span>ad:<span class=w> </span><span class=o>[</span><span class=m>0</span>,<span class=w> </span><span class=m>74</span><span class=o>]</span>,<span class=w> </span>gq:<span class=w> </span><span class=m>301</span><span class=o>}</span>
<span class=o>{</span>id:<span class=w> </span><span class=m>10342734968077036194</span>,<span class=w> </span>sample:<span class=w> </span><span class=s2>&quot;HG001&quot;</span>,<span class=w> </span>gt:<span class=w> </span><span class=m>2</span>,<span class=w> </span>ps:<span class=w> </span>null,<span class=w> </span>dp:<span class=w> </span><span class=m>820</span>,<span class=w> </span>adall:<span class=w> </span><span class=o>[</span><span class=m>125</span>,<span class=w> </span><span class=m>383</span><span class=o>]</span>,<span class=w> </span>ad:<span class=w> </span><span class=o>[</span><span class=m>0</span>,<span class=w> </span><span class=m>70</span><span class=o>]</span>,<span class=w> </span>gq:<span class=w> </span><span class=m>339</span><span class=o>}</span>
<span class=o>{</span>id:<span class=w> </span><span class=m>890514037559296207</span>,<span class=w> </span>sample:<span class=w> </span><span class=s2>&quot;HG001&quot;</span>,<span class=w> </span>gt:<span class=w> </span><span class=m>1</span>,<span class=w> </span>ps:<span class=w> </span>null,<span class=w> </span>dp:<span class=w> </span><span class=m>760</span>,<span class=w> </span>adall:<span class=w> </span><span class=o>[</span><span class=m>161</span>,<span class=w> </span><span class=m>142</span><span class=o>]</span>,<span class=w> </span>ad:<span class=w> </span><span class=o>[</span><span class=m>25</span>,<span class=w> </span><span class=m>37</span><span class=o>]</span>,<span class=w> </span>gq:<span class=w> </span><span class=m>147</span><span class=o>}</span>
</code></pre></div></p> </details> <h2 id=structuration-of-data>Structuration of data<a class=headerlink href=#structuration-of-data title="Permanent link">&para;</a></h2> <h3 id=merge-all-variant>Merge all variant<a class=headerlink href=#merge-all-variant title="Permanent link">&para;</a></h3> <p>We can now aggregate all variant present in our dataset to perform this operation we use divide to conquer merge method by generate temporary file. By default, file are written in <code>/tmp</code> but you can control where these files are written by set <code>TMPDIR</code>, <code>TEMP</code> or <code>TMP</code> directory.</p> <div class=highlight><pre><span></span><code>variantplaner<span class=w> </span>-t<span class=w> </span><span class=m>8</span><span class=w> </span>struct<span class=w> </span>-i<span class=w> </span>variants/*.parquet<span class=w> </span>--<span class=w> </span>variants<span class=w> </span>-o<span class=w> </span>uniq_variants/
</code></pre></div> <p>Directory <code>uniq_variants</code> contains all unique variants present in dataset split by chromosome, <code>--</code> after last input path are mandatory.</p> <h3 id=genotypes-structuration>Genotypes structuration<a class=headerlink href=#genotypes-structuration title="Permanent link">&para;</a></h3> <h3 id=by-samples>By samples<a class=headerlink href=#by-samples title="Permanent link">&para;</a></h3> <p>This structurations data is already down in vcf2parquet step check content of <code>genotypes/samples</code>: <div class=highlight><pre><span></span><code>➜<span class=w> </span>ls<span class=w> </span>genotypes/samples
HG001.parquet<span class=w>  </span>HG002.parquet<span class=w>  </span>HG003.parquet<span class=w>  </span>HG004.parquet<span class=w>  </span>HG006.parquet<span class=w>  </span>HG007.parquet
</code></pre></div></p> <h3 id=by-variants>By variants<a class=headerlink href=#by-variants title="Permanent link">&para;</a></h3> <p>Here, we'll organize the genotypes information by variants to make it easier to find samples where a variant is present or not.</p> <div class=highlight><pre><span></span><code>mkdir<span class=w> </span>-p<span class=w> </span>genotypes/variants/
variantplaner<span class=w> </span>-t<span class=w> </span><span class=m>8</span><span class=w> </span>struct<span class=w> </span>-i<span class=w> </span>genotypes/samples/*.parquet<span class=w> </span>--<span class=w> </span>genotypes<span class=w> </span>-p<span class=w> </span>genotypes/variants
</code></pre></div> <p>All genotypes information are split in <a href=https://duckdb.org/docs/data/partitioning/hive_partitioning>hive like structure</a> to optimize request on data, <code>--</code> after last input path are mandatory.</p> <h3 id=compute-transmission-mode>Compute transmission mode<a class=headerlink href=#compute-transmission-mode title="Permanent link">&para;</a></h3> <p>If you are working with families, <code>variantplaner</code> can calculate the modes of transmission of the variants.</p> <p>For these step, we need to concatenate all genotypes of a AshkenazimTrio in one parquet sample.</p> <div class=highlight><pre><span></span><code>pqrs<span class=w> </span>merge<span class=w> </span>-i<span class=w> </span>genotypes/samples/HG002.parquet<span class=w> </span>genotypes/samples/HG003.parquet<span class=w> </span>genotypes/samples/HG004.parquet<span class=w> </span>-o<span class=w> </span>genotypes/samples/AshkenazimTrio.parquet
mkdir<span class=w> </span>-p<span class=w> </span>genotypes/transmissions/
variantplaner<span class=w> </span>transmission<span class=w> </span>-g<span class=w> </span>genotypes/samples/AshkenazimTrio.parquet<span class=w> </span>-c<span class=w> </span>HG002<span class=w> </span>-m<span class=w> </span>HG003<span class=w> </span>-f<span class=w> </span>HG004<span class=w> </span>-o<span class=w> </span>genotypes/transmissions/AshkenazimTrio.parquet
</code></pre></div> <p><code>-c</code> parameter is use for index sample, <code>-m</code> parameter is use for mother sample, <code>-f</code> parameter is use for father sample only the index sample is mandatory if mother sample or father sample isn't present command work, you could also use a pedigree file with parameter <code>-p</code>.</p> <p>If in your genotypes file you have multiple child or multiple trio you could repeat option <code>--child</code>, <code>--father</code> and <code>--mother</code>.</p> <p>Example <div class=highlight><pre><span></span><code>variantplaner<span class=w> </span>transmission<span class=w> </span>-g<span class=w> </span>genotypes.parquet<span class=w> </span>-o<span class=w> </span>transmissions.parquet<span class=w> </span><span class=se>\</span>
-c<span class=w> </span>child1<span class=w> </span>-m<span class=w> </span>mother<span class=w> </span>-f<span class=w> </span>father<span class=w> </span><span class=se>\</span>
-c<span class=w> </span>child2<span class=w> </span>-m<span class=w> </span>mother<span class=w> </span>-f<span class=w> </span>father<span class=w> </span><span class=se>\</span>
-c<span class=w> </span>child3
</code></pre></div></p> <p>If all child have same parent you not need to repeat mother and father parameter.</p> <details> <summary>transmission parquet file content</summary> <div class=highlight><pre><span></span><code>{id: 73978904708120583, sample: &quot;HG002&quot;, index_dp: 956, index_gq: 477, index_adall: [0, 252], index_ad: [80, 530], index_gt: 2, index_ps: null, mother_dp: 898, mother_gq: 547, mother_adall: [141, 142], mother_ad: [253, 236], mother_gt: 1, mother_ps: null, father_dp: 935, father_gq: 542, father_adall: [146, 124], father_ad: [246, 233], father_gt: 1, father_ps: null, origin: &quot;#&quot;&quot;&quot;}
{id: 73978997049917444, sample: &quot;HG002&quot;, index_dp: 971, index_gq: 469, index_adall: [0, 253], index_ad: [77, 533], index_gt: 2, index_ps: null, mother_dp: 875, mother_gq: 559, mother_adall: [140, 135], mother_ad: [238, 234], mother_gt: 1, mother_ps: null, father_dp: 881, father_gq: 551, father_adall: [137, 128], father_ad: [222, 226], father_gt: 1, father_ps: null, origin: &quot;#&quot;&quot;&quot;}
{id: 73981277677551621, sample: &quot;HG002&quot;, index_dp: 1016, index_gq: 483, index_adall: [0, 268], index_ad: [87, 571], index_gt: 2, index_ps: null, mother_dp: 877, mother_gq: 551, mother_adall: [140, 137], mother_ad: [236, 237], mother_gt: 1, mother_ps: null, father_dp: 933, father_gq: 594, father_adall: [150, 149], father_ad: [252, 261], father_gt: 1, father_ps: null, origin: &quot;#&quot;&quot;&quot;}
{id: 73981928365096967, sample: &quot;HG002&quot;, index_dp: 999, index_gq: 595, index_adall: [0, 296], index_ad: [114, 551], index_gt: 2, index_ps: null, mother_dp: 969, mother_gq: 551, mother_adall: [174, 133], mother_ad: [299, 233], mother_gt: 1, mother_ps: null, father_dp: 948, father_gq: 555, father_adall: [131, 129], father_ad: [271, 275], father_gt: 1, father_ps: null, origin: &quot;#&quot;&quot;&quot;}
{id: 73982503890714628, sample: &quot;HG002&quot;, index_dp: 1048, index_gq: 559, index_adall: [1, 317], index_ad: [59, 554], index_gt: 2, index_ps: null, mother_dp: 939, mother_gq: 554, mother_adall: [166, 145], mother_ad: [273, 231], mother_gt: 1, mother_ps: null, father_dp: 1017, father_gq: 564, father_adall: [139, 173], father_ad: [278, 271], father_gt: 1, father_ps: null, origin: &quot;#&quot;&quot;&quot;}
</code></pre></div> </details> <p>Parquet transmissions file contains column all genotypes information with suffix <code>_index</code>, <code>_mother</code> and <code>_father</code> plus a <code>origin</code> column</p> <p>Origin column contains a string with 3 character:</p> <div class=highlight><pre><span></span><code>#~&quot;
││└ ASCII_value_of(father genotype + 33)
│└─ ASCII_value_of(mother genotype + 33)
└── ASCII_value_of(index genotype  + 33)
</code></pre></div> <p>In this example case, variants is homozygotes in index, mother information is missing, variants is heterozygotes in father.</p> <p>Maximal genotype value is 92, which corresponds to the character <code>}</code>, <code>~</code> match with value 93, this value also mean unknow genotype.</p> <h2 id=add-annotations>Add annotations<a class=headerlink href=#add-annotations title="Permanent link">&para;</a></h2> <p>To work on your variant, you probably need variants annotations.</p> <p>Because many's annotation vcf file header does not contain information on contigs, you should create file <code>grch38.92.csv</code> with this content: <div class=highlight><pre><span></span><code>contig,length
chr1,248956422
chr10,133797422
chr11,135086622
chr12,133275309
chr13,114364328
chr14,107043718
chr15,101991189
chr16,90338345
chr17,83257441
chr18,80373285
chr19,58617616
chr2,242193529
chr20,64444167
chr21,46709983
chr22,50818468
chr3,198295559
chr4,190214555
chr5,181538259
chr6,170805979
chr7,159345973
chr8,145138636
chr9,138394717
chrMT,16569
chrX,156040895
chrY,57227415
</code></pre></div></p> <p>Header name are mandatory. Take care of chromosome order are the same between your vcf header and <code>grch38.92.csv</code> and name are same.</p> <h3 id=snpeff-annotations>Snpeff annotations<a class=headerlink href=#snpeff-annotations title="Permanent link">&para;</a></h3> <p>First convert your unique variants in parquet format (<code>variants.parquet</code>) in vcf: <div class=highlight><pre><span></span><code>variantplaner<span class=w> </span>-t<span class=w> </span><span class=m>8</span><span class=w> </span>parquet2vcf<span class=w> </span>-i<span class=w> </span>variants.parquet<span class=w> </span>-o<span class=w> </span>variants.vcf
</code></pre></div></p> <p><code>parquet2vcf</code> subcommand have many more options but we didn't need it now.</p> <p>Next annotate this <code>variants.vcf</code> <a href=https://pcingola.github.io/SnpEff/ >with snpeff</a>, we assume you generate a file call <code>variants.snpeff.vcf</code>.</p> <p>To convert annotated vcf in parquet, keep 'ANN' info column and rename vcf id column in snpeff_id you can run: <div class=highlight><pre><span></span><code>mkdir<span class=w> </span>-p<span class=w> </span>annotations
variantplaner<span class=w> </span>-t<span class=w> </span><span class=m>8</span><span class=w> </span>vcf2parquet<span class=w> </span>-c<span class=w> </span>grch38.92.csv<span class=w> </span>-i<span class=w> </span>variants.snpeff.vcf<span class=w> </span>annotations<span class=w> </span>-o<span class=w> </span>annotations/snpeff.parquet<span class=w> </span>vcf<span class=w> </span>-i<span class=w> </span>ANN<span class=w> </span>-r<span class=w> </span>snpeff_id
</code></pre></div></p> <p>If you didn't set any value of option <code>-i</code> in vcf subsubcommand all info column are keep.</p> <h3 id=clinvar-annotations>Clinvar annotations<a class=headerlink href=#clinvar-annotations title="Permanent link">&para;</a></h3> <p>Download last clinvar version:</p> <div class=highlight><pre><span></span><code>mkdir<span class=w> </span>-p<span class=w> </span>annotations
curl<span class=w> </span>https://ftp.ncbi.nlm.nih.gov/pub/clinvar/vcf_GRCh38/clinvar.vcf.gz<span class=w> </span>&gt;<span class=w> </span>annotations/clinvar.vcf.gz
</code></pre></div> <p>Convert clinvar vcf file in parquet file:</p> <div class=highlight><pre><span></span><code>variantplaner<span class=w> </span>vcf2parquet<span class=w> </span>-c<span class=w> </span>grch38.92.csv<span class=w> </span>-i<span class=w> </span>annotations/clinvar.vcf<span class=w> </span>annotations<span class=w> </span>-o<span class=w> </span>annotations/clinvar.parquet<span class=w> </span>-r<span class=w> </span>clinvar_id
</code></pre></div> <p>Parquet file produce contains many columns:</p> <ul> <li>clinvar_id: content of vcf id column if option <code>-r</code> is not set, column name is <code>vid</code></li> <li>id: variantplaner id</li> <li>All INFO filed</li> </ul> <p>Annotations subcommand try to make match between vcf info type and parquet type.</p> <details> <summary><code>annotations/clinvar.parquet</code> file content</summary> <div class=highlight><pre><span></span><code>pqrs<span class=w> </span>head<span class=w> </span>annotations/clinvar.parquet
<span class=o>{</span>clinvar_id:<span class=w> </span><span class=m>2205837</span>,<span class=w> </span>id:<span class=w> </span><span class=m>11650605831284591550</span>,<span class=w> </span>AF_ESP:<span class=w> </span>null,<span class=w> </span>AF_EXAC:<span class=w> </span>null,<span class=w> </span>AF_TGP:<span class=w> </span>null,<span class=w> </span>ALLELEID:<span class=w> </span><span class=m>2193183</span>,<span class=w> </span>CLNDN:<span class=w> </span><span class=o>[</span><span class=s2>&quot;Inborn_genetic_diseases&quot;</span><span class=o>]</span>,<span class=w> </span>CLNDNINCL:<span class=w> </span>null,<span class=w> </span>CLNDISDB:<span class=w> </span><span class=o>[</span><span class=s2>&quot;MeSH:D030342&quot;</span>,<span class=w> </span><span class=s2>&quot;MedGen:C0950123&quot;</span><span class=o>]</span>,<span class=w> </span>CLNDISDBINCL:<span class=w> </span>null,<span class=w> </span>CLNHGVS:<span class=w> </span><span class=o>[</span><span class=s2>&quot;NC_000001.11:g.69134A&gt;G&quot;</span><span class=o>]</span>,<span class=w> </span>CLNREVSTAT:<span class=w> </span><span class=o>[</span><span class=s2>&quot;criteria_provided&quot;</span>,<span class=w> </span><span class=s2>&quot;_single_submitter&quot;</span><span class=o>]</span>,<span class=w> </span>CLNSIG:<span class=w> </span><span class=o>[</span><span class=s2>&quot;Likely_benign&quot;</span><span class=o>]</span>,<span class=w> </span>CLNSIGCONF:<span class=w> </span>null,<span class=w> </span>CLNSIGINCL:<span class=w> </span>null,<span class=w> </span>CLNVC:<span class=w> </span><span class=s2>&quot;single_nucleotide_variant&quot;</span>,<span class=w> </span>CLNVCSO:<span class=w> </span><span class=s2>&quot;SO:0001483&quot;</span>,<span class=w> </span>CLNVI:<span class=w> </span>null,<span class=w> </span>DBVARID:<span class=w> </span>null,<span class=w> </span>GENEINFO:<span class=w> </span><span class=s2>&quot;OR4F5:79501&quot;</span>,<span class=w> </span>MC:<span class=w> </span><span class=o>[</span><span class=s2>&quot;SO:0001583|missense_variant&quot;</span><span class=o>]</span>,<span class=w> </span>ORIGIN:<span class=w> </span><span class=o>[</span><span class=s2>&quot;1&quot;</span><span class=o>]</span>,<span class=w> </span>RS:<span class=w> </span>null<span class=o>}</span>
<span class=o>{</span>clinvar_id:<span class=w> </span><span class=m>2252161</span>,<span class=w> </span>id:<span class=w> </span><span class=m>2295086632353399847</span>,<span class=w> </span>AF_ESP:<span class=w> </span>null,<span class=w> </span>AF_EXAC:<span class=w> </span>null,<span class=w> </span>AF_TGP:<span class=w> </span>null,<span class=w> </span>ALLELEID:<span class=w> </span><span class=m>2238986</span>,<span class=w> </span>CLNDN:<span class=w> </span><span class=o>[</span><span class=s2>&quot;Inborn_genetic_diseases&quot;</span><span class=o>]</span>,<span class=w> </span>CLNDNINCL:<span class=w> </span>null,<span class=w> </span>CLNDISDB:<span class=w> </span><span class=o>[</span><span class=s2>&quot;MeSH:D030342&quot;</span>,<span class=w> </span><span class=s2>&quot;MedGen:C0950123&quot;</span><span class=o>]</span>,<span class=w> </span>CLNDISDBINCL:<span class=w> </span>null,<span class=w> </span>CLNHGVS:<span class=w> </span><span class=o>[</span><span class=s2>&quot;NC_000001.11:g.69581C&gt;G&quot;</span><span class=o>]</span>,<span class=w> </span>CLNREVSTAT:<span class=w> </span><span class=o>[</span><span class=s2>&quot;criteria_provided&quot;</span>,<span class=w> </span><span class=s2>&quot;_single_submitter&quot;</span><span class=o>]</span>,<span class=w> </span>CLNSIG:<span class=w> </span><span class=o>[</span><span class=s2>&quot;Uncertain_significance&quot;</span><span class=o>]</span>,<span class=w> </span>CLNSIGCONF:<span class=w> </span>null,<span class=w> </span>CLNSIGINCL:<span class=w> </span>null,<span class=w> </span>CLNVC:<span class=w> </span><span class=s2>&quot;single_nucleotide_variant&quot;</span>,<span class=w> </span>CLNVCSO:<span class=w> </span><span class=s2>&quot;SO:0001483&quot;</span>,<span class=w> </span>CLNVI:<span class=w> </span>null,<span class=w> </span>DBVARID:<span class=w> </span>null,<span class=w> </span>GENEINFO:<span class=w> </span><span class=s2>&quot;OR4F5:79501&quot;</span>,<span class=w> </span>MC:<span class=w> </span><span class=o>[</span><span class=s2>&quot;SO:0001583|missense_variant&quot;</span><span class=o>]</span>,<span class=w> </span>ORIGIN:<span class=w> </span><span class=o>[</span><span class=s2>&quot;1&quot;</span><span class=o>]</span>,<span class=w> </span>RS:<span class=w> </span>null<span class=o>}</span>
<span class=o>{</span>clinvar_id:<span class=w> </span><span class=m>2396347</span>,<span class=w> </span>id:<span class=w> </span><span class=m>11033100074712141168</span>,<span class=w> </span>AF_ESP:<span class=w> </span>null,<span class=w> </span>AF_EXAC:<span class=w> </span>null,<span class=w> </span>AF_TGP:<span class=w> </span>null,<span class=w> </span>ALLELEID:<span class=w> </span><span class=m>2386655</span>,<span class=w> </span>CLNDN:<span class=w> </span><span class=o>[</span><span class=s2>&quot;Inborn_genetic_diseases&quot;</span><span class=o>]</span>,<span class=w> </span>CLNDNINCL:<span class=w> </span>null,<span class=w> </span>CLNDISDB:<span class=w> </span><span class=o>[</span><span class=s2>&quot;MeSH:D030342&quot;</span>,<span class=w> </span><span class=s2>&quot;MedGen:C0950123&quot;</span><span class=o>]</span>,<span class=w> </span>CLNDISDBINCL:<span class=w> </span>null,<span class=w> </span>CLNHGVS:<span class=w> </span><span class=o>[</span><span class=s2>&quot;NC_000001.11:g.69682G&gt;A&quot;</span><span class=o>]</span>,<span class=w> </span>CLNREVSTAT:<span class=w> </span><span class=o>[</span><span class=s2>&quot;criteria_provided&quot;</span>,<span class=w> </span><span class=s2>&quot;_single_submitter&quot;</span><span class=o>]</span>,<span class=w> </span>CLNSIG:<span class=w> </span><span class=o>[</span><span class=s2>&quot;Uncertain_significance&quot;</span><span class=o>]</span>,<span class=w> </span>CLNSIGCONF:<span class=w> </span>null,<span class=w> </span>CLNSIGINCL:<span class=w> </span>null,<span class=w> </span>CLNVC:<span class=w> </span><span class=s2>&quot;single_nucleotide_variant&quot;</span>,<span class=w> </span>CLNVCSO:<span class=w> </span><span class=s2>&quot;SO:0001483&quot;</span>,<span class=w> </span>CLNVI:<span class=w> </span>null,<span class=w> </span>DBVARID:<span class=w> </span>null,<span class=w> </span>GENEINFO:<span class=w> </span><span class=s2>&quot;OR4F5:79501&quot;</span>,<span class=w> </span>MC:<span class=w> </span><span class=o>[</span><span class=s2>&quot;SO:0001583|missense_variant&quot;</span><span class=o>]</span>,<span class=w> </span>ORIGIN:<span class=w> </span><span class=o>[</span><span class=s2>&quot;1&quot;</span><span class=o>]</span>,<span class=w> </span>RS:<span class=w> </span>null<span class=o>}</span>
<span class=o>{</span>clinvar_id:<span class=w> </span><span class=m>2288999</span>,<span class=w> </span>id:<span class=w> </span><span class=m>10487392163259126218</span>,<span class=w> </span>AF_ESP:<span class=w> </span>null,<span class=w> </span>AF_EXAC:<span class=w> </span>null,<span class=w> </span>AF_TGP:<span class=w> </span>null,<span class=w> </span>ALLELEID:<span class=w> </span><span class=m>2278803</span>,<span class=w> </span>CLNDN:<span class=w> </span><span class=o>[</span><span class=s2>&quot;Inborn_genetic_diseases&quot;</span><span class=o>]</span>,<span class=w> </span>CLNDNINCL:<span class=w> </span>null,<span class=w> </span>CLNDISDB:<span class=w> </span><span class=o>[</span><span class=s2>&quot;MeSH:D030342&quot;</span>,<span class=w> </span><span class=s2>&quot;MedGen:C0950123&quot;</span><span class=o>]</span>,<span class=w> </span>CLNDISDBINCL:<span class=w> </span>null,<span class=w> </span>CLNHGVS:<span class=w> </span><span class=o>[</span><span class=s2>&quot;NC_000001.11:g.69769T&gt;C&quot;</span><span class=o>]</span>,<span class=w> </span>CLNREVSTAT:<span class=w> </span><span class=o>[</span><span class=s2>&quot;criteria_provided&quot;</span>,<span class=w> </span><span class=s2>&quot;_single_submitter&quot;</span><span class=o>]</span>,<span class=w> </span>CLNSIG:<span class=w> </span><span class=o>[</span><span class=s2>&quot;Uncertain_significance&quot;</span><span class=o>]</span>,<span class=w> </span>CLNSIGCONF:<span class=w> </span>null,<span class=w> </span>CLNSIGINCL:<span class=w> </span>null,<span class=w> </span>CLNVC:<span class=w> </span><span class=s2>&quot;single_nucleotide_variant&quot;</span>,<span class=w> </span>CLNVCSO:<span class=w> </span><span class=s2>&quot;SO:0001483&quot;</span>,<span class=w> </span>CLNVI:<span class=w> </span>null,<span class=w> </span>DBVARID:<span class=w> </span>null,<span class=w> </span>GENEINFO:<span class=w> </span><span class=s2>&quot;OR4F5:79501&quot;</span>,<span class=w> </span>MC:<span class=w> </span><span class=o>[</span><span class=s2>&quot;SO:0001583|missense_variant&quot;</span><span class=o>]</span>,<span class=w> </span>ORIGIN:<span class=w> </span><span class=o>[</span><span class=s2>&quot;1&quot;</span><span class=o>]</span>,<span class=w> </span>RS:<span class=w> </span>null<span class=o>}</span>
<span class=o>{</span>clinvar_id:<span class=w> </span><span class=m>2351346</span>,<span class=w> </span>id:<span class=w> </span><span class=m>5356120651941363990</span>,<span class=w> </span>AF_ESP:<span class=w> </span>null,<span class=w> </span>AF_EXAC:<span class=w> </span>null,<span class=w> </span>AF_TGP:<span class=w> </span>null,<span class=w> </span>ALLELEID:<span class=w> </span><span class=m>2333177</span>,<span class=w> </span>CLNDN:<span class=w> </span><span class=o>[</span><span class=s2>&quot;Inborn_genetic_diseases&quot;</span><span class=o>]</span>,<span class=w> </span>CLNDNINCL:<span class=w> </span>null,<span class=w> </span>CLNDISDB:<span class=w> </span><span class=o>[</span><span class=s2>&quot;MeSH:D030342&quot;</span>,<span class=w> </span><span class=s2>&quot;MedGen:C0950123&quot;</span><span class=o>]</span>,<span class=w> </span>CLNDISDBINCL:<span class=w> </span>null,<span class=w> </span>CLNHGVS:<span class=w> </span><span class=o>[</span><span class=s2>&quot;NC_000001.11:g.69995G&gt;C&quot;</span><span class=o>]</span>,<span class=w> </span>CLNREVSTAT:<span class=w> </span><span class=o>[</span><span class=s2>&quot;criteria_provided&quot;</span>,<span class=w> </span><span class=s2>&quot;_single_submitter&quot;</span><span class=o>]</span>,<span class=w> </span>CLNSIG:<span class=w> </span><span class=o>[</span><span class=s2>&quot;Uncertain_significance&quot;</span><span class=o>]</span>,<span class=w> </span>CLNSIGCONF:<span class=w> </span>null,<span class=w> </span>CLNSIGINCL:<span class=w> </span>null,<span class=w> </span>CLNVC:<span class=w> </span><span class=s2>&quot;single_nucleotide_variant&quot;</span>,<span class=w> </span>CLNVCSO:<span class=w> </span><span class=s2>&quot;SO:0001483&quot;</span>,<span class=w> </span>CLNVI:<span class=w> </span>null,<span class=w> </span>DBVARID:<span class=w> </span>null,<span class=w> </span>GENEINFO:<span class=w> </span><span class=s2>&quot;OR4F5:79501&quot;</span>,<span class=w> </span>MC:<span class=w> </span><span class=o>[</span><span class=s2>&quot;SO:0001583|missense_variant&quot;</span><span class=o>]</span>,<span class=w> </span>ORIGIN:<span class=w> </span><span class=o>[</span><span class=s2>&quot;1&quot;</span><span class=o>]</span>,<span class=w> </span>RS:<span class=w> </span>null<span class=o>}</span>
</code></pre></div> </details> <p>With option of subcommand vcf <code>-i</code> you can select which column are included in parquet file For example command: <div class=highlight><pre><span></span><code>variantplaner<span class=w> </span>vcf2parquet<span class=w> </span>-c<span class=w> </span>grch38.92.csv<span class=w> </span>-i<span class=w> </span>annotations/clinvar.vcf<span class=w> </span>annotations<span class=w> </span>-o<span class=w> </span>annotations/clinvar.parquet<span class=w> </span>-r<span class=w> </span>clinvar_id<span class=w> </span>-i<span class=w> </span>ALLELEID<span class=w> </span>-i<span class=w> </span>CLNDN<span class=w> </span>-i<span class=w> </span>AF_ESP<span class=w> </span>-i<span class=w> </span>GENEINFO
</code></pre></div></p> <p>Produce a <code>annotations/clinvar.parquet</code> with columns:</p> <ul> <li>clinvar_id</li> <li>id</li> <li>ALLELEID</li> <li>CLNDN</li> </ul> <details> <summary><code>annotations/clinvar.parquet</code> file content</summary> <div class=highlight><pre><span></span><code>➜<span class=w> </span>pqrs<span class=w> </span>head<span class=w> </span>annotations/clinvar.parquet
<span class=o>{</span>clinvar_id:<span class=w> </span><span class=m>2205837</span>,<span class=w> </span>id:<span class=w> </span><span class=m>11650605831284591550</span>,<span class=w> </span>ALLELEID:<span class=w> </span><span class=m>2193183</span>,<span class=w> </span>CLNDN:<span class=w> </span><span class=o>[</span><span class=s2>&quot;Inborn_genetic_diseases&quot;</span><span class=o>]}</span>
<span class=o>{</span>clinvar_id:<span class=w> </span><span class=m>2252161</span>,<span class=w> </span>id:<span class=w> </span><span class=m>2295086632353399847</span>,<span class=w> </span>ALLELEID:<span class=w> </span><span class=m>2238986</span>,<span class=w> </span>CLNDN:<span class=w> </span><span class=o>[</span><span class=s2>&quot;Inborn_genetic_diseases&quot;</span><span class=o>]}</span>
<span class=o>{</span>clinvar_id:<span class=w> </span><span class=m>2396347</span>,<span class=w> </span>id:<span class=w> </span><span class=m>11033100074712141168</span>,<span class=w> </span>ALLELEID:<span class=w> </span><span class=m>2386655</span>,<span class=w> </span>CLNDN:<span class=w> </span><span class=o>[</span><span class=s2>&quot;Inborn_genetic_diseases&quot;</span><span class=o>]}</span>
<span class=o>{</span>clinvar_id:<span class=w> </span><span class=m>2288999</span>,<span class=w> </span>id:<span class=w> </span><span class=m>10487392163259126218</span>,<span class=w> </span>ALLELEID:<span class=w> </span><span class=m>2278803</span>,<span class=w> </span>CLNDN:<span class=w> </span><span class=o>[</span><span class=s2>&quot;Inborn_genetic_diseases&quot;</span><span class=o>]}</span>
<span class=o>{</span>clinvar_id:<span class=w> </span><span class=m>2351346</span>,<span class=w> </span>id:<span class=w> </span><span class=m>5356120651941363990</span>,<span class=w> </span>ALLELEID:<span class=w> </span><span class=m>2333177</span>,<span class=w> </span>CLNDN:<span class=w> </span><span class=o>[</span><span class=s2>&quot;Inborn_genetic_diseases&quot;</span><span class=o>]}</span>
</code></pre></div> </details> <h3 id=split-snpeff-annotations-in-parquet-column>Split snpeff annotations in parquet column<a class=headerlink href=#split-snpeff-annotations-in-parquet-column title="Permanent link">&para;</a></h3> <p>After annotate your vcf with snpeff in <code>ann</code> mode and write result in file <code>snpeff_annotations.vcf</code> you could run a script similar to this to generate <code>snpeff_annotations.parquet</code>: <div class=highlight><pre><span></span><code><span class=kn>from</span><span class=w> </span><span class=nn>variantplaner</span><span class=w> </span><span class=kn>import</span> <span class=n>Vcf</span>

<span class=n>vcf</span> <span class=o>=</span> <span class=n>Vcf</span><span class=p>()</span>
<span class=k>try</span><span class=p>:</span>
    <span class=n>vcf</span><span class=o>.</span><span class=n>from_path</span><span class=p>(</span><span class=s2>&quot;snpeff_annotations.vcf&quot;</span><span class=p>,</span> <span class=s2>&quot;grch38.92.csv&quot;</span><span class=p>)</span>
<span class=k>except</span> <span class=n>variantplaner</span><span class=o>.</span><span class=n>exception</span><span class=o>.</span><span class=n>NotAVCFError</span><span class=p>:</span>
    <span class=nb>print</span><span class=p>(</span><span class=s2>&quot;snpeff_annotations.vcf seems to have error&quot;</span><span class=p>)</span>
    <span class=k>return</span> <span class=mi>1</span>
<span class=k>except</span> <span class=n>variantplaner</span><span class=o>.</span><span class=n>exception</span><span class=o>.</span><span class=n>NoContigsLengthInformationError</span><span class=p>:</span>
    <span class=nb>print</span><span class=p>(</span><span class=s2>&quot;snpeff_annotations.vcf header seems not contain contig information&quot;</span><span class=p>)</span>
    <span class=k>return</span> <span class=mi>2</span>

<span class=n>lf</span> <span class=o>=</span> <span class=n>vcf</span><span class=o>.</span><span class=n>lf</span><span class=o>.</span><span class=n>with_columns</span><span class=p>(</span><span class=n>vcf</span><span class=o>.</span><span class=n>header</span><span class=o>.</span><span class=n>info_parser</span><span class=p>())</span>
<span class=n>lf</span> <span class=o>=</span> <span class=n>lf</span><span class=o>.</span><span class=n>drop</span><span class=p>([</span><span class=s2>&quot;chr&quot;</span><span class=p>,</span> <span class=s2>&quot;pos&quot;</span><span class=p>,</span> <span class=s2>&quot;ref&quot;</span><span class=p>,</span> <span class=s2>&quot;alt&quot;</span><span class=p>,</span> <span class=s2>&quot;filter&quot;</span><span class=p>,</span> <span class=s2>&quot;qual&quot;</span><span class=p>,</span> <span class=s2>&quot;info&quot;</span><span class=p>])</span>
<span class=n>lf</span> <span class=o>=</span> <span class=n>lf</span><span class=o>.</span><span class=n>rename</span><span class=p>({</span><span class=s2>&quot;vid&quot;</span><span class=p>:</span> <span class=s2>&quot;id&quot;</span><span class=p>})</span>
<span class=n>lf</span> <span class=o>=</span> <span class=n>lf</span><span class=o>.</span><span class=n>explode</span><span class=p>(</span><span class=s2>&quot;ANN&quot;</span><span class=p>)</span>
<span class=n>lf</span> <span class=o>=</span> <span class=n>lf</span><span class=o>.</span><span class=n>cast</span><span class=p>({</span><span class=s2>&quot;id&quot;</span><span class=p>:</span> <span class=n>polars</span><span class=o>.</span><span class=n>UInt64</span><span class=p>})</span>

<span class=n>lf</span> <span class=o>=</span> <span class=n>lf</span><span class=o>.</span><span class=n>with_columns</span><span class=p>(</span>
    <span class=p>[</span>
        <span class=n>polars</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;ANN&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>str</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=s2>&quot;|&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>cast</span><span class=p>(</span><span class=n>polars</span><span class=o>.</span><span class=n>List</span><span class=p>(</span><span class=n>polars</span><span class=o>.</span><span class=n>Utf8</span><span class=p>()))</span>
        <span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;ann&quot;</span><span class=p>),</span>
    <span class=p>]</span>
<span class=p>)</span><span class=o>.</span><span class=n>drop</span><span class=p>(</span><span class=s2>&quot;ANN&quot;</span><span class=p>)</span>

<span class=n>lf</span> <span class=o>=</span> <span class=n>lf</span><span class=o>.</span><span class=n>with_columns</span><span class=p>(</span>
    <span class=p>[</span>
        <span class=n>polars</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;ann&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>list</span><span class=o>.</span><span class=n>get</span><span class=p>(</span><span class=mi>1</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;effect&quot;</span><span class=p>),</span>
        <span class=n>polars</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;ann&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>list</span><span class=o>.</span><span class=n>get</span><span class=p>(</span><span class=mi>2</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;impact&quot;</span><span class=p>),</span>
        <span class=n>polars</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;ann&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>list</span><span class=o>.</span><span class=n>get</span><span class=p>(</span><span class=mi>3</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;gene&quot;</span><span class=p>),</span>
        <span class=n>polars</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;ann&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>list</span><span class=o>.</span><span class=n>get</span><span class=p>(</span><span class=mi>4</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;geneid&quot;</span><span class=p>),</span>
        <span class=n>polars</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;ann&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>list</span><span class=o>.</span><span class=n>get</span><span class=p>(</span><span class=mi>5</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;feature&quot;</span><span class=p>),</span>
        <span class=n>polars</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;ann&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>list</span><span class=o>.</span><span class=n>get</span><span class=p>(</span><span class=mi>6</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;feature_id&quot;</span><span class=p>),</span>
        <span class=n>polars</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;ann&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>list</span><span class=o>.</span><span class=n>get</span><span class=p>(</span><span class=mi>7</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;bio_type&quot;</span><span class=p>),</span>
        <span class=n>polars</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;ann&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>list</span><span class=o>.</span><span class=n>get</span><span class=p>(</span><span class=mi>8</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;rank&quot;</span><span class=p>),</span>
        <span class=n>polars</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;ann&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>list</span><span class=o>.</span><span class=n>get</span><span class=p>(</span><span class=mi>9</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;hgvs_c&quot;</span><span class=p>),</span>
        <span class=n>polars</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;ann&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>list</span><span class=o>.</span><span class=n>get</span><span class=p>(</span><span class=mi>10</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;hgvs_p&quot;</span><span class=p>),</span>
        <span class=n>polars</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;ann&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>list</span><span class=o>.</span><span class=n>get</span><span class=p>(</span><span class=mi>11</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;cdna_pos&quot;</span><span class=p>),</span>
        <span class=n>polars</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;ann&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>list</span><span class=o>.</span><span class=n>get</span><span class=p>(</span><span class=mi>12</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;cdna_len&quot;</span><span class=p>),</span>
        <span class=n>polars</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;ann&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>list</span><span class=o>.</span><span class=n>get</span><span class=p>(</span><span class=mi>13</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;cds_pos&quot;</span><span class=p>),</span>
        <span class=n>polars</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;ann&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>list</span><span class=o>.</span><span class=n>get</span><span class=p>(</span><span class=mi>14</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;cds_len&quot;</span><span class=p>),</span>
        <span class=n>polars</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;ann&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>list</span><span class=o>.</span><span class=n>get</span><span class=p>(</span><span class=mi>15</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;aa_pos&quot;</span><span class=p>),</span>
    <span class=p>]</span>
<span class=p>)</span><span class=o>.</span><span class=n>drop</span><span class=p>(</span><span class=s2>&quot;ann&quot;</span><span class=p>)</span>

<span class=n>lf</span><span class=o>.</span><span class=n>sink_parquet</span><span class=p>(</span><span class=s2>&quot;snpeff_annotations.parquet&quot;</span><span class=p>)</span>
</code></pre></div></p> <h2 id=querying>Querying<a class=headerlink href=#querying title="Permanent link">&para;</a></h2> <p>You can use any tool or software library supporting the parquet format to use the files generated by <code>variantplaner</code>.</p> <p>We show you how to use files with <a href=https://crates.io/crates/polars-cli>polars-cli</a> and <a href=https://duckdb.org/ >duckdb</a>.</p> <h3 id=polars-cli>polars-cli<a class=headerlink href=#polars-cli title="Permanent link">&para;</a></h3> <h4 id=count-variants>Count variants<a class=headerlink href=#count-variants title="Permanent link">&para;</a></h4> <div class=highlight><pre><span></span><code><span class=err>〉</span><span class=k>select</span><span class=w> </span><span class=k>count</span><span class=p>(</span><span class=o>*</span><span class=p>)</span><span class=w> </span><span class=k>from</span><span class=w> </span><span class=n>read_parquet</span><span class=p>(</span><span class=s1>&#39;variants.parquet&#39;</span><span class=p>);</span>
<span class=err>┌─────────┐</span>
<span class=err>│</span><span class=w> </span><span class=k>count</span><span class=w>   </span><span class=err>│</span>
<span class=err>│</span><span class=w> </span><span class=c1>---     │</span>
<span class=err>│</span><span class=w> </span><span class=n>u32</span><span class=w>     </span><span class=err>│</span>
<span class=err>╞═════════╡</span>
<span class=err>│</span><span class=w> </span><span class=mi>7852699</span><span class=w> </span><span class=err>│</span>
<span class=err>└─────────┘</span>
</code></pre></div> <details> <summary>check result with <code>pqrs</code></summary> <p>We can check we have same result with pqrs</p> <div class=highlight><pre><span></span><code>➜<span class=w> </span>pqrs<span class=w> </span>rowcount<span class=w> </span>variants.parquet
File<span class=w> </span>Name:<span class=w> </span>variants.parquet:<span class=w> </span><span class=m>7852699</span><span class=w> </span>rows
</code></pre></div> </details> <h4 id=filter-variants-from-annotations>Filter variants from annotations:<a class=headerlink href=#filter-variants-from-annotations title="Permanent link">&para;</a></h4> <p>Get all variant with a AF_ESP upper than 0.9999</p> <div class=highlight><pre><span></span><code><span class=err>〉</span><span class=k>select</span><span class=w> </span><span class=n>chr</span><span class=p>,</span><span class=w> </span><span class=n>pos</span><span class=p>,</span><span class=w> </span><span class=k>ref</span><span class=p>,</span><span class=w> </span><span class=n>alt</span><span class=p>,</span><span class=w> </span><span class=n>AF_ESP</span><span class=w> </span><span class=k>from</span><span class=w> </span><span class=n>read_parquet</span><span class=p>(</span><span class=s1>&#39;variants.parquet&#39;</span><span class=p>)</span><span class=w> </span><span class=k>as</span><span class=w> </span><span class=n>v</span><span class=w> </span><span class=k>left</span><span class=w> </span><span class=k>join</span><span class=w> </span><span class=n>read_parquet</span><span class=p>(</span><span class=s1>&#39;annotations/clinvar.parquet&#39;</span><span class=p>)</span><span class=w> </span><span class=k>as</span><span class=w> </span><span class=k>c</span><span class=w> </span><span class=k>on</span><span class=w> </span><span class=k>c</span><span class=p>.</span><span class=n>id</span><span class=o>=</span><span class=n>v</span><span class=p>.</span><span class=n>id</span><span class=w> </span><span class=k>where</span><span class=w> </span><span class=n>AF_ESP</span><span class=o>&gt;</span><span class=mi>0</span><span class=p>.</span><span class=mi>9999</span><span class=p>;</span>
<span class=err>┌─────┬──────────┬─────┬─────┬─────────┐</span>
<span class=err>│</span><span class=w> </span><span class=n>chr</span><span class=w> </span><span class=err>┆</span><span class=w> </span><span class=n>pos</span><span class=w>      </span><span class=err>┆</span><span class=w> </span><span class=k>ref</span><span class=w> </span><span class=err>┆</span><span class=w> </span><span class=n>alt</span><span class=w> </span><span class=err>┆</span><span class=w> </span><span class=n>AF_ESP</span><span class=w>  </span><span class=err>│</span>
<span class=err>│</span><span class=w> </span><span class=c1>--- ┆ ---      ┆ --- ┆ --- ┆ ---     │</span>
<span class=err>│</span><span class=w> </span><span class=n>u8</span><span class=w>  </span><span class=err>┆</span><span class=w> </span><span class=n>u64</span><span class=w>      </span><span class=err>┆</span><span class=w> </span><span class=n>str</span><span class=w> </span><span class=err>┆</span><span class=w> </span><span class=n>str</span><span class=w> </span><span class=err>┆</span><span class=w> </span><span class=n>f64</span><span class=w>     </span><span class=err>│</span>
<span class=err>╞═════╪══════════╪═════╪═════╪═════════╡</span>
<span class=err>│</span><span class=w> </span><span class=mi>10</span><span class=w>  </span><span class=err>┆</span><span class=w> </span><span class=mi>16901372</span><span class=w> </span><span class=err>┆</span><span class=w> </span><span class=k>G</span><span class=w>   </span><span class=err>┆</span><span class=w> </span><span class=k>C</span><span class=w>   </span><span class=err>┆</span><span class=w> </span><span class=mi>0</span><span class=p>.</span><span class=mi>99992</span><span class=w> </span><span class=err>│</span>
<span class=err>│</span><span class=w> </span><span class=mi>11</span><span class=w>  </span><span class=err>┆</span><span class=w> </span><span class=mi>78121030</span><span class=w> </span><span class=err>┆</span><span class=w> </span><span class=n>T</span><span class=w>   </span><span class=err>┆</span><span class=w> </span><span class=n>A</span><span class=w>   </span><span class=err>┆</span><span class=w> </span><span class=mi>0</span><span class=p>.</span><span class=mi>99992</span><span class=w> </span><span class=err>│</span>
<span class=err>└─────┴──────────┴─────┴─────┴─────────┘</span>
</code></pre></div> <h4 id=get-sample-have-variant>Get sample have variant<a class=headerlink href=#get-sample-have-variant title="Permanent link">&para;</a></h4> <p>Get all variant and sample with GENEINFO equal to 'SAMD11:148398'</p> <div class=highlight><pre><span></span><code><span class=err>〉</span><span class=k>select</span><span class=w> </span><span class=k>distinct</span><span class=w> </span><span class=n>chr</span><span class=p>,</span><span class=w> </span><span class=n>pos</span><span class=p>,</span><span class=w> </span><span class=k>ref</span><span class=p>,</span><span class=w> </span><span class=n>alt</span><span class=p>,</span><span class=w> </span><span class=n>sample</span><span class=w> </span><span class=k>from</span><span class=w> </span><span class=n>read_parquet</span><span class=p>(</span><span class=s1>&#39;variants.parquet&#39;</span><span class=p>)</span><span class=w> </span><span class=k>as</span><span class=w> </span><span class=n>v</span><span class=w> </span><span class=k>left</span><span class=w> </span><span class=k>join</span><span class=w> </span><span class=n>read_parquet</span><span class=p>(</span><span class=s1>&#39;genotypes/samples/*&#39;</span><span class=p>)</span><span class=w> </span><span class=k>as</span><span class=w> </span><span class=k>g</span><span class=w> </span><span class=k>on</span><span class=w> </span><span class=n>v</span><span class=p>.</span><span class=n>id</span><span class=o>=</span><span class=k>g</span><span class=p>.</span><span class=n>id</span><span class=w> </span><span class=k>left</span><span class=w> </span><span class=k>join</span><span class=w> </span><span class=n>read_parquet</span><span class=p>(</span><span class=s1>&#39;annotations/clinvar.parquet&#39;</span><span class=p>)</span><span class=w> </span><span class=k>as</span><span class=w> </span><span class=n>a</span><span class=w> </span><span class=k>on</span><span class=w> </span><span class=n>v</span><span class=p>.</span><span class=n>id</span><span class=o>=</span><span class=n>a</span><span class=p>.</span><span class=n>id</span><span class=w> </span><span class=k>WHERE</span><span class=w> </span><span class=n>GENEINFO</span><span class=o>=</span><span class=s1>&#39;SAMD11:148398&#39;</span><span class=p>;</span>
<span class=err>┌─────┬────────┬─────┬─────┬────────┐</span>
<span class=err>│</span><span class=w> </span><span class=n>chr</span><span class=w> </span><span class=err>┆</span><span class=w> </span><span class=n>pos</span><span class=w>    </span><span class=err>┆</span><span class=w> </span><span class=k>ref</span><span class=w> </span><span class=err>┆</span><span class=w> </span><span class=n>alt</span><span class=w> </span><span class=err>┆</span><span class=w> </span><span class=n>sample</span><span class=w> </span><span class=err>│</span>
<span class=err>│</span><span class=w> </span><span class=c1>--- ┆ ---    ┆ --- ┆ --- ┆ ---    │</span>
<span class=err>│</span><span class=w> </span><span class=n>u8</span><span class=w>  </span><span class=err>┆</span><span class=w> </span><span class=n>u64</span><span class=w>    </span><span class=err>┆</span><span class=w> </span><span class=n>str</span><span class=w> </span><span class=err>┆</span><span class=w> </span><span class=n>str</span><span class=w> </span><span class=err>┆</span><span class=w> </span><span class=n>str</span><span class=w>    </span><span class=err>│</span>
<span class=err>╞═════╪════════╪═════╪═════╪════════╡</span>
<span class=err>│</span><span class=w> </span><span class=mi>1</span><span class=w>   </span><span class=err>┆</span><span class=w> </span><span class=mi>942451</span><span class=w> </span><span class=err>┆</span><span class=w> </span><span class=n>T</span><span class=w>   </span><span class=err>┆</span><span class=w> </span><span class=k>C</span><span class=w>   </span><span class=err>┆</span><span class=w> </span><span class=n>HG003</span><span class=w>  </span><span class=err>│</span>
<span class=err>│</span><span class=w> </span><span class=mi>1</span><span class=w>   </span><span class=err>┆</span><span class=w> </span><span class=mi>942451</span><span class=w> </span><span class=err>┆</span><span class=w> </span><span class=n>T</span><span class=w>   </span><span class=err>┆</span><span class=w> </span><span class=k>C</span><span class=w>   </span><span class=err>┆</span><span class=w> </span><span class=n>HG001</span><span class=w>  </span><span class=err>│</span>
<span class=err>│</span><span class=w> </span><span class=mi>1</span><span class=w>   </span><span class=err>┆</span><span class=w> </span><span class=mi>942451</span><span class=w> </span><span class=err>┆</span><span class=w> </span><span class=n>T</span><span class=w>   </span><span class=err>┆</span><span class=w> </span><span class=k>C</span><span class=w>   </span><span class=err>┆</span><span class=w> </span><span class=n>HG007</span><span class=w>  </span><span class=err>│</span>
<span class=err>│</span><span class=w> </span><span class=mi>1</span><span class=w>   </span><span class=err>┆</span><span class=w> </span><span class=mi>942451</span><span class=w> </span><span class=err>┆</span><span class=w> </span><span class=n>T</span><span class=w>   </span><span class=err>┆</span><span class=w> </span><span class=k>C</span><span class=w>   </span><span class=err>┆</span><span class=w> </span><span class=n>HG004</span><span class=w>  </span><span class=err>│</span>
<span class=err>│</span><span class=w> </span><span class=err>…</span><span class=w>   </span><span class=err>┆</span><span class=w> </span><span class=err>…</span><span class=w>      </span><span class=err>┆</span><span class=w> </span><span class=err>…</span><span class=w>   </span><span class=err>┆</span><span class=w> </span><span class=err>…</span><span class=w>   </span><span class=err>┆</span><span class=w> </span><span class=err>…</span><span class=w>      </span><span class=err>│</span>
<span class=err>│</span><span class=w> </span><span class=mi>1</span><span class=w>   </span><span class=err>┆</span><span class=w> </span><span class=mi>942934</span><span class=w> </span><span class=err>┆</span><span class=w> </span><span class=k>G</span><span class=w>   </span><span class=err>┆</span><span class=w> </span><span class=k>C</span><span class=w>   </span><span class=err>┆</span><span class=w> </span><span class=n>HG003</span><span class=w>  </span><span class=err>│</span>
<span class=err>│</span><span class=w> </span><span class=mi>1</span><span class=w>   </span><span class=err>┆</span><span class=w> </span><span class=mi>943937</span><span class=w> </span><span class=err>┆</span><span class=w> </span><span class=k>C</span><span class=w>   </span><span class=err>┆</span><span class=w> </span><span class=n>T</span><span class=w>   </span><span class=err>┆</span><span class=w> </span><span class=n>HG004</span><span class=w>  </span><span class=err>│</span>
<span class=err>│</span><span class=w> </span><span class=mi>1</span><span class=w>   </span><span class=err>┆</span><span class=w> </span><span class=mi>942451</span><span class=w> </span><span class=err>┆</span><span class=w> </span><span class=n>T</span><span class=w>   </span><span class=err>┆</span><span class=w> </span><span class=k>C</span><span class=w>   </span><span class=err>┆</span><span class=w> </span><span class=n>HG002</span><span class=w>  </span><span class=err>│</span>
<span class=err>│</span><span class=w> </span><span class=mi>1</span><span class=w>   </span><span class=err>┆</span><span class=w> </span><span class=mi>942451</span><span class=w> </span><span class=err>┆</span><span class=w> </span><span class=n>T</span><span class=w>   </span><span class=err>┆</span><span class=w> </span><span class=k>C</span><span class=w>   </span><span class=err>┆</span><span class=w> </span><span class=n>HG006</span><span class=w>  </span><span class=err>│</span>
<span class=err>└─────┴────────┴─────┴─────┴────────┘</span>
</code></pre></div> <h3 id=duckdb>duckdb<a class=headerlink href=#duckdb title="Permanent link">&para;</a></h3> <h4 id=count-variants_1>Count variants<a class=headerlink href=#count-variants_1 title="Permanent link">&para;</a></h4> <div class=highlight><pre><span></span><code><span class=n>D</span><span class=w> </span><span class=k>select</span><span class=w> </span><span class=k>count</span><span class=p>(</span><span class=o>*</span><span class=p>)</span><span class=w> </span><span class=k>from</span><span class=w> </span><span class=n>read_parquet</span><span class=p>(</span><span class=s1>&#39;variants.parquet&#39;</span><span class=p>);</span>
<span class=err>┌──────────────┐</span>
<span class=err>│</span><span class=w> </span><span class=n>count_star</span><span class=p>()</span><span class=w> </span><span class=err>│</span>
<span class=err>│</span><span class=w>    </span><span class=n>int64</span><span class=w>     </span><span class=err>│</span>
<span class=err>├──────────────┤</span>
<span class=err>│</span><span class=w>      </span><span class=mi>7852699</span><span class=w> </span><span class=err>│</span>
<span class=err>└──────────────┘</span>
</code></pre></div> <details> <summary>check result with <code>pqrs</code></summary> <p>We can check we have same result with pqrs</p> <div class=highlight><pre><span></span><code>➜<span class=w> </span>pqrs<span class=w> </span>rowcount<span class=w> </span>variants.parquet
File<span class=w> </span>Name:<span class=w> </span>variants.parquet:<span class=w> </span><span class=m>7852699</span><span class=w> </span>rows
</code></pre></div> </details> <h4 id=filter-variants-from-annotations_1>Filter variants from annotations:<a class=headerlink href=#filter-variants-from-annotations_1 title="Permanent link">&para;</a></h4> <p>Get all variant with a AF_ESP upper than 0.9999</p> <div class=highlight><pre><span></span><code><span class=n>D</span><span class=w> </span><span class=k>select</span><span class=w> </span><span class=n>chr</span><span class=p>,</span><span class=w> </span><span class=n>pos</span><span class=p>,</span><span class=w> </span><span class=k>ref</span><span class=p>,</span><span class=w> </span><span class=n>alt</span><span class=p>,</span><span class=w> </span><span class=n>AF_ESP</span><span class=w> </span><span class=k>from</span><span class=w> </span><span class=n>read_parquet</span><span class=p>(</span><span class=s1>&#39;variants.parquet&#39;</span><span class=p>)</span><span class=w> </span><span class=k>as</span><span class=w> </span><span class=n>v</span><span class=w> </span><span class=k>left</span><span class=w> </span><span class=k>join</span><span class=w> </span><span class=n>read_parquet</span><span class=p>(</span><span class=s1>&#39;annotations/clinvar.parquet&#39;</span><span class=p>)</span><span class=w> </span><span class=k>as</span><span class=w> </span><span class=k>c</span><span class=w> </span><span class=k>on</span><span class=w> </span><span class=k>c</span><span class=p>.</span><span class=n>id</span><span class=o>=</span><span class=n>v</span><span class=p>.</span><span class=n>id</span><span class=w> </span><span class=k>where</span><span class=w> </span><span class=n>AF_ESP</span><span class=o>&gt;</span><span class=mi>0</span><span class=p>.</span><span class=mi>9999</span><span class=p>;</span>
<span class=err>┌───────┬──────────┬─────────┬─────────┬─────────┐</span>
<span class=err>│</span><span class=w>  </span><span class=n>chr</span><span class=w>  </span><span class=err>│</span><span class=w>   </span><span class=n>pos</span><span class=w>    </span><span class=err>│</span><span class=w>   </span><span class=k>ref</span><span class=w>   </span><span class=err>│</span><span class=w>   </span><span class=n>alt</span><span class=w>   </span><span class=err>│</span><span class=w> </span><span class=n>AF_ESP</span><span class=w>  </span><span class=err>│</span>
<span class=err>│</span><span class=w> </span><span class=n>uint8</span><span class=w> </span><span class=err>│</span><span class=w>  </span><span class=n>uint64</span><span class=w>  </span><span class=err>│</span><span class=w> </span><span class=nb>varchar</span><span class=w> </span><span class=err>│</span><span class=w> </span><span class=nb>varchar</span><span class=w> </span><span class=err>│</span><span class=w> </span><span class=n>double</span><span class=w>  </span><span class=err>│</span>
<span class=err>├───────┼──────────┼─────────┼─────────┼─────────┤</span>
<span class=err>│</span><span class=w>    </span><span class=mi>10</span><span class=w> </span><span class=err>│</span><span class=w> </span><span class=mi>16901372</span><span class=w> </span><span class=err>│</span><span class=w> </span><span class=k>G</span><span class=w>       </span><span class=err>│</span><span class=w> </span><span class=k>C</span><span class=w>       </span><span class=err>│</span><span class=w> </span><span class=mi>0</span><span class=p>.</span><span class=mi>99992</span><span class=w> </span><span class=err>│</span>
<span class=err>│</span><span class=w>    </span><span class=mi>11</span><span class=w> </span><span class=err>│</span><span class=w> </span><span class=mi>78121030</span><span class=w> </span><span class=err>│</span><span class=w> </span><span class=n>T</span><span class=w>       </span><span class=err>│</span><span class=w> </span><span class=n>A</span><span class=w>       </span><span class=err>│</span><span class=w> </span><span class=mi>0</span><span class=p>.</span><span class=mi>99992</span><span class=w> </span><span class=err>│</span>
<span class=err>└───────┴──────────┴─────────┴─────────┴─────────┘</span>
</code></pre></div> <h4 id=get-sample-have-variant_1>Get sample have variant<a class=headerlink href=#get-sample-have-variant_1 title="Permanent link">&para;</a></h4> <p>Get all variant and sample with GENEINFO equal to 'SAMD11:148398'</p> <div class=highlight><pre><span></span><code><span class=n>D</span><span class=w> </span><span class=k>select</span><span class=w> </span><span class=k>distinct</span><span class=w> </span><span class=n>chr</span><span class=p>,</span><span class=w> </span><span class=n>pos</span><span class=p>,</span><span class=w> </span><span class=k>ref</span><span class=p>,</span><span class=w> </span><span class=n>alt</span><span class=p>,</span><span class=w> </span><span class=n>sample</span><span class=w> </span><span class=k>from</span><span class=w> </span><span class=n>read_parquet</span><span class=p>(</span><span class=s1>&#39;variants.parquet&#39;</span><span class=p>)</span><span class=w> </span><span class=k>as</span><span class=w> </span><span class=n>v</span><span class=w> </span><span class=k>left</span><span class=w> </span><span class=k>join</span><span class=w> </span><span class=n>read_parquet</span><span class=p>(</span><span class=s1>&#39;genotypes/samples/*&#39;</span><span class=p>)</span><span class=w> </span><span class=k>as</span><span class=w> </span><span class=k>g</span><span class=w> </span><span class=k>on</span><span class=w> </span><span class=n>v</span><span class=p>.</span><span class=n>id</span><span class=o>=</span><span class=k>g</span><span class=p>.</span><span class=n>id</span><span class=w> </span><span class=k>left</span><span class=w> </span><span class=k>join</span><span class=w> </span><span class=n>read_parquet</span><span class=p>(</span><span class=s1>&#39;annotations/clinvar.parquet&#39;</span><span class=p>)</span><span class=w> </span><span class=k>as</span><span class=w> </span><span class=n>a</span><span class=w> </span><span class=k>on</span><span class=w> </span><span class=n>v</span><span class=p>.</span><span class=n>id</span><span class=o>=</span><span class=n>a</span><span class=p>.</span><span class=n>id</span><span class=w> </span><span class=k>WHERE</span><span class=w> </span><span class=n>GENEINFO</span><span class=o>=</span><span class=s1>&#39;SAMD11:148398&#39;</span><span class=p>;</span>
<span class=err>┌───────┬────────┬─────────┬─────────┬─────────┐</span>
<span class=err>│</span><span class=w>  </span><span class=n>chr</span><span class=w>  </span><span class=err>│</span><span class=w>  </span><span class=n>pos</span><span class=w>   </span><span class=err>│</span><span class=w>   </span><span class=k>ref</span><span class=w>   </span><span class=err>│</span><span class=w>   </span><span class=n>alt</span><span class=w>   </span><span class=err>│</span><span class=w> </span><span class=n>sample</span><span class=w>  </span><span class=err>│</span>
<span class=err>│</span><span class=w> </span><span class=n>uint8</span><span class=w> </span><span class=err>│</span><span class=w> </span><span class=n>uint64</span><span class=w> </span><span class=err>│</span><span class=w> </span><span class=nb>varchar</span><span class=w> </span><span class=err>│</span><span class=w> </span><span class=nb>varchar</span><span class=w> </span><span class=err>│</span><span class=w> </span><span class=nb>varchar</span><span class=w> </span><span class=err>│</span>
<span class=err>├───────┼────────┼─────────┼─────────┼─────────┤</span>
<span class=err>│</span><span class=w>     </span><span class=mi>1</span><span class=w> </span><span class=err>│</span><span class=w> </span><span class=mi>942451</span><span class=w> </span><span class=err>│</span><span class=w> </span><span class=n>T</span><span class=w>       </span><span class=err>│</span><span class=w> </span><span class=k>C</span><span class=w>       </span><span class=err>│</span><span class=w> </span><span class=n>HG002</span><span class=w>   </span><span class=err>│</span>
<span class=err>│</span><span class=w>     </span><span class=mi>1</span><span class=w> </span><span class=err>│</span><span class=w> </span><span class=mi>942934</span><span class=w> </span><span class=err>│</span><span class=w> </span><span class=k>G</span><span class=w>       </span><span class=err>│</span><span class=w> </span><span class=k>C</span><span class=w>       </span><span class=err>│</span><span class=w> </span><span class=n>HG002</span><span class=w>   </span><span class=err>│</span>
<span class=err>│</span><span class=w>     </span><span class=mi>1</span><span class=w> </span><span class=err>│</span><span class=w> </span><span class=mi>942451</span><span class=w> </span><span class=err>│</span><span class=w> </span><span class=n>T</span><span class=w>       </span><span class=err>│</span><span class=w> </span><span class=k>C</span><span class=w>       </span><span class=err>│</span><span class=w> </span><span class=n>HG003</span><span class=w>   </span><span class=err>│</span>
<span class=err>│</span><span class=w>     </span><span class=mi>1</span><span class=w> </span><span class=err>│</span><span class=w> </span><span class=mi>942934</span><span class=w> </span><span class=err>│</span><span class=w> </span><span class=k>G</span><span class=w>       </span><span class=err>│</span><span class=w> </span><span class=k>C</span><span class=w>       </span><span class=err>│</span><span class=w> </span><span class=n>HG003</span><span class=w>   </span><span class=err>│</span>
<span class=err>│</span><span class=w>     </span><span class=mi>1</span><span class=w> </span><span class=err>│</span><span class=w> </span><span class=mi>942451</span><span class=w> </span><span class=err>│</span><span class=w> </span><span class=n>T</span><span class=w>       </span><span class=err>│</span><span class=w> </span><span class=k>C</span><span class=w>       </span><span class=err>│</span><span class=w> </span><span class=n>HG007</span><span class=w>   </span><span class=err>│</span>
<span class=err>│</span><span class=w>     </span><span class=mi>1</span><span class=w> </span><span class=err>│</span><span class=w> </span><span class=mi>943937</span><span class=w> </span><span class=err>│</span><span class=w> </span><span class=k>C</span><span class=w>       </span><span class=err>│</span><span class=w> </span><span class=n>T</span><span class=w>       </span><span class=err>│</span><span class=w> </span><span class=n>HG007</span><span class=w>   </span><span class=err>│</span>
<span class=err>│</span><span class=w>     </span><span class=mi>1</span><span class=w> </span><span class=err>│</span><span class=w> </span><span class=mi>942451</span><span class=w> </span><span class=err>│</span><span class=w> </span><span class=n>T</span><span class=w>       </span><span class=err>│</span><span class=w> </span><span class=k>C</span><span class=w>       </span><span class=err>│</span><span class=w> </span><span class=n>HG001</span><span class=w>   </span><span class=err>│</span>
<span class=err>│</span><span class=w>     </span><span class=mi>1</span><span class=w> </span><span class=err>│</span><span class=w> </span><span class=mi>942451</span><span class=w> </span><span class=err>│</span><span class=w> </span><span class=n>T</span><span class=w>       </span><span class=err>│</span><span class=w> </span><span class=k>C</span><span class=w>       </span><span class=err>│</span><span class=w> </span><span class=n>HG004</span><span class=w>   </span><span class=err>│</span>
<span class=err>│</span><span class=w>     </span><span class=mi>1</span><span class=w> </span><span class=err>│</span><span class=w> </span><span class=mi>943937</span><span class=w> </span><span class=err>│</span><span class=w> </span><span class=k>C</span><span class=w>       </span><span class=err>│</span><span class=w> </span><span class=n>T</span><span class=w>       </span><span class=err>│</span><span class=w> </span><span class=n>HG004</span><span class=w>   </span><span class=err>│</span>
<span class=err>│</span><span class=w>     </span><span class=mi>1</span><span class=w> </span><span class=err>│</span><span class=w> </span><span class=mi>942451</span><span class=w> </span><span class=err>│</span><span class=w> </span><span class=n>T</span><span class=w>       </span><span class=err>│</span><span class=w> </span><span class=k>C</span><span class=w>       </span><span class=err>│</span><span class=w> </span><span class=n>HG006</span><span class=w>   </span><span class=err>│</span>
<span class=err>├───────┴────────┴─────────┴─────────┴─────────┤</span>
<span class=err>│</span><span class=w> </span><span class=mi>10</span><span class=w> </span><span class=k>rows</span><span class=w>                            </span><span class=mi>5</span><span class=w> </span><span class=n>columns</span><span class=w> </span><span class=err>│</span>
<span class=err>└──────────────────────────────────────────────┘</span>
</code></pre></div> <h3 id=use-genotype-partition>Use genotype partition<a class=headerlink href=#use-genotype-partition title="Permanent link">&para;</a></h3> <p>In this example, I'll show how I interact with the data structures created by variantplaner.</p> <h4 id=import>Import<a class=headerlink href=#import title="Permanent link">&para;</a></h4> <div class=highlight><pre><span></span><code><span class=kn>import</span><span class=w> </span><span class=nn>duckdb</span>
<span class=kn>import</span><span class=w> </span><span class=nn>polars</span>
<span class=kn>import</span><span class=w> </span><span class=nn>variantplaner</span>
</code></pre></div> <h4 id=get-variants>Get variants<a class=headerlink href=#get-variants title="Permanent link">&para;</a></h4> <div class=highlight><pre><span></span><code><span class=n>query</span> <span class=o>=</span> <span class=sa>f</span><span class=s2>&quot;&quot;&quot;</span>
<span class=s2>SELECT</span>
<span class=s2>  *</span>
<span class=s2>FROM</span>
<span class=s2>  read_parquet(&#39;variants.parquet&#39;) as v</span>
<span class=s2>WHERE</span>
<span class=s2>  v.chr == &#39;19&#39;</span>
<span class=s2>&quot;&quot;&quot;</span>

<span class=n>variants</span> <span class=o>=</span> <span class=n>duckdb</span><span class=o>.</span><span class=n>query</span><span class=p>(</span><span class=n>query</span><span class=p>)</span><span class=o>.</span><span class=n>pl</span><span class=p>()</span>
</code></pre></div> <h4 id=add-annotations_1>Add annotations<a class=headerlink href=#add-annotations_1 title="Permanent link">&para;</a></h4> <div class=highlight><pre><span></span><code><span class=n>query</span> <span class=o>=</span> <span class=sa>f</span><span class=s2>&quot;&quot;&quot;</span>
<span class=s2>SELECT</span>
<span class=s2>  v.*, c.CLNSIG</span>
<span class=s2>FROM</span>
<span class=s2>  variants as v</span>
<span class=s2>  JOIN</span>
<span class=s2>  read_parquet(&#39;annotations/clinvar.parquet&#39;) as c</span>
<span class=s2>  ON</span>
<span class=s2>  v.id == c.id</span>
<span class=s2>WHERE</span>
<span class=s2>  c.CLNSIG LIKE &#39;%Patho%&#39;</span>
<span class=s2>&quot;&quot;&quot;</span>

<span class=n>annotations</span> <span class=o>=</span> <span class=n>duckdb</span><span class=o>.</span><span class=n>query</span><span class=p>(</span><span class=n>query</span><span class=p>)</span><span class=o>.</span><span class=n>pl</span><span class=p>()</span>
</code></pre></div> <h4 id=add-genotypes>Add genotypes<a class=headerlink href=#add-genotypes title="Permanent link">&para;</a></h4> <div class=highlight><pre><span></span><code><span class=k>def</span><span class=w> </span><span class=nf>worker</span><span class=p>(</span><span class=n>name_data</span><span class=p>:</span> <span class=p>(</span><span class=nb>str</span><span class=p>,</span> <span class=n>polars</span><span class=o>.</span><span class=n>DataFrame</span><span class=p>))</span> <span class=o>-&gt;</span> <span class=n>polars</span><span class=o>.</span><span class=n>DataFrame</span><span class=p>:</span>
<span class=w>    </span><span class=sd>&quot;&quot;&quot;Request genotype homozygote variant.&quot;&quot;&quot;</span>
    <span class=n>name</span><span class=p>,</span> <span class=n>data</span> <span class=o>=</span> <span class=n>name_data</span>
    <span class=n>query</span> <span class=o>=</span> <span class=sa>f</span><span class=s2>&quot;&quot;&quot;</span>
<span class=s2>SELECT</span>
<span class=s2>  data.*, g.*,</span>
<span class=s2>FROM</span>
<span class=s2>  data</span>
<span class=s2>  JOIN</span>
<span class=s2>  read_parquet(&#39;genotypes/variants/id_part=</span><span class=si>{</span><span class=n>name</span><span class=si>}</span><span class=s2>/0.parquet&#39;) as g ON data.id = g.id</span>
<span class=s2>WHERE</span>
<span class=s2>  g.gt == 2</span>
<span class=s2>&quot;&quot;&quot;</span>
    <span class=n>df</span> <span class=o>=</span> <span class=n>duckdb</span><span class=o>.</span><span class=n>query</span><span class=p>(</span><span class=n>query</span><span class=p>)</span><span class=o>.</span><span class=n>pl</span><span class=p>()</span>
    <span class=k>return</span> <span class=n>df</span>


<span class=n>annotations</span> <span class=o>=</span> <span class=n>variantplaner</span><span class=o>.</span><span class=n>normalization</span><span class=o>.</span><span class=n>add_id_part</span><span class=p>(</span><span class=n>annotations</span><span class=p>)</span>

<span class=n>all_genotypes</span> <span class=o>=</span> <span class=p>[]</span>

<span class=k>for</span> <span class=n>data</span> <span class=ow>in</span> <span class=nb>map</span><span class=p>(</span><span class=n>worker</span><span class=p>,</span> <span class=n>annotations</span><span class=o>.</span><span class=n>group_by</span><span class=p>(</span><span class=n>by</span><span class=o>=</span><span class=s2>&quot;id_part&quot;</span><span class=p>)):</span>
    <span class=k>if</span> <span class=n>data</span> <span class=ow>is</span> <span class=ow>not</span> <span class=kc>None</span><span class=p>:</span>
        <span class=n>all_genotypes</span><span class=o>.</span><span class=n>append</span><span class=p>(</span><span class=n>data</span><span class=p>)</span>

<span class=n>genotypes</span> <span class=o>=</span> <span class=n>polars</span><span class=o>.</span><span class=n>concat</span><span class=p>(</span><span class=n>all_genotypes</span><span class=p>)</span>
</code></pre></div> <p>genotypes is polars.DataFrame with pathogene homozygote variants in chromosome 19.</p> <aside class=md-source-file> <span class=md-source-file__fact> <span class=md-icon title="Last update"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M21 13.1c-.1 0-.3.1-.4.2l-1 1 2.1 2.1 1-1c.2-.2.2-.6 0-.8l-1.3-1.3c-.1-.1-.2-.2-.4-.2m-1.9 1.8-6.1 6V23h2.1l6.1-6.1zM12.5 7v5.2l4 2.4-1 1L11 13V7zM11 21.9c-5.1-.5-9-4.8-9-9.9C2 6.5 6.5 2 12 2c5.3 0 9.6 4.1 10 9.3-.3-.1-.6-.2-1-.2s-.7.1-1 .2C19.6 7.2 16.2 4 12 4c-4.4 0-8 3.6-8 8 0 4.1 3.1 7.5 7.1 7.9l-.1.2z"/></svg> </span> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-timeago" title="June 18, 2025 14:50:17 UTC"><span class=timeago datetime=2025-06-18T14:50:17+00:00 locale=en></span></span><span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-iso_date" title="June 18, 2025 14:50:17 UTC">2025-06-18</span> </span> <span class=md-source-file__fact> <span class=md-icon title=Created> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M14.47 15.08 11 13V7h1.5v5.25l3.08 1.83c-.41.28-.79.62-1.11 1m-1.39 4.84c-.36.05-.71.08-1.08.08-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8c0 .37-.03.72-.08 1.08.69.1 1.33.32 1.92.64.1-.56.16-1.13.16-1.72 0-5.5-4.5-10-10-10S2 6.5 2 12s4.47 10 10 10c.59 0 1.16-.06 1.72-.16-.32-.59-.54-1.23-.64-1.92M18 15v3h-3v2h3v3h2v-3h3v-2h-3v-3z"/></svg> </span> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-timeago" title="April 27, 2023 11:42:05 UTC"><span class=timeago datetime=2023-04-27T11:42:05+00:00 locale=en></span></span><span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-iso_date" title="April 27, 2023 11:42:05 UTC">2023-04-27</span> </span> </aside> <!-- Giscus --> <!-- https://squidfunk.github.io/mkdocs-material/setup/adding-a-comment-system/#giscus-integration --> <div id=feedback style="display: none;"> <h2 id=__comments>Feedback</h2> <script src=https://giscus.app/client.js data-repo=SeqOIA-IT/variantplaner data-repo-id="REPLACE WITH REPOSITORY NODE ID" data-category=Documentation data-category-id="REPLACE WITH DISCUSSION CATEGORY ID" data-mapping=pathname data-strict=1 data-reactions-enabled=0 data-emit-metadata=0 data-input-position=top data-theme=preferred_color_scheme data-lang=en data-loading=lazy crossorigin=anonymous async>
    </script> <!-- Synchronize Giscus theme with palette --> <script>
        var giscus = document.querySelector("script[src*=giscus]")

        // Set palette on initial load
        var palette = __md_get("__palette")
        if (palette && typeof palette.color === "object") {
            var theme = palette.color.scheme === "slate"
                ? "transparent_dark"
                : "light"

            // Instruct Giscus to set theme
            giscus.setAttribute("data-theme", theme)
        }

        // Register event handlers after documented loaded
        document.addEventListener("DOMContentLoaded", function() {
            var ref = document.querySelector("[data-md-component=palette]")
            ref.addEventListener("change", function() {
                var palette = __md_get("__palette")
                if (palette && typeof palette.color === "object") {
                    var theme = palette.color.scheme === "slate"
                        ? "transparent_dark"
                        : "light"

                    // Instruct Giscus to change theme
                    var frame = document.querySelector(".giscus-frame")
                    frame.contentWindow.postMessage(
                        { giscus: { setConfig: { theme } } },
                        "https://giscus.app"
                    )
                }
            })
        })
    </script> </div> </article> </div> <script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script> </div> <button type=button class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg> Back to top </button> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=.. class="md-footer__link md-footer__link--prev" aria-label="Previous: Overview"> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg> </div> <div class=md-footer__title> <span class=md-footer__direction> Previous </span> <div class=md-ellipsis> Overview </div> </div> </a> <a href=../changelog/ class="md-footer__link md-footer__link--next" aria-label="Next: Changelog"> <div class=md-footer__title> <span class=md-footer__direction> Next </span> <div class=md-ellipsis> Changelog </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> <div class=md-copyright__highlight> Copyright &copy; 2025 GCS SeqOIA </div> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://github.com/SeqOIA-IT target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8M97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg> </a> <a href=https://gitter.im/variantplaner/community target=_blank rel=noopener title=gitter.im class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 384 512"><!-- Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M66.4 322.5H16V0h50.4zM166.9 76.1h-50.4V512h50.4zm100.6 0h-50.4V512h50.4zM368 76h-50.4v247H368z"/></svg> </a> <a href=https://pypi.org/project/variantplaner/ target=_blank rel=noopener title=pypi.org class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.8 200.5c-7.7-30.9-22.3-54.2-53.4-54.2h-40.1v47.4c0 36.8-31.2 67.8-66.8 67.8H172.7c-29.2 0-53.4 25-53.4 54.3v101.8c0 29 25.2 46 53.4 54.3 33.8 9.9 66.3 11.7 106.8 0 26.9-7.8 53.4-23.5 53.4-54.3v-40.7H226.2v-13.6h160.2c31.1 0 42.6-21.7 53.4-54.2 11.2-33.5 10.7-65.7 0-108.6M286.2 404c11.1 0 20.1 9.1 20.1 20.3 0 11.3-9 20.4-20.1 20.4-11 0-20.1-9.2-20.1-20.4.1-11.3 9.1-20.3 20.1-20.3M167.8 248.1h106.8c29.7 0 53.4-24.5 53.4-54.3V91.9c0-29-24.4-50.7-53.4-55.6-35.8-5.9-74.7-5.6-106.8.1-45.2 8-53.4 24.7-53.4 55.6v40.7h106.9v13.6h-147c-31.1 0-58.3 18.7-66.8 54.2-9.8 40.7-10.2 66.1 0 108.6 7.6 31.6 25.7 54.2 56.8 54.2H101v-48.8c0-35.3 30.5-66.4 66.8-66.4m-6.7-142.6c-11.1 0-20.1-9.1-20.1-20.3.1-11.3 9-20.4 20.1-20.4 11 0 20.1 9.2 20.1 20.4s-9 20.3-20.1 20.3"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": "..", "features": ["announce.dismiss", "content.action.edit", "content.action.view", "content.code.annotate", "content.code.copy", "content.tooltips", "navigation.footer", "navigation.instant.preview", "navigation.path", "navigation.sections", "navigation.tabs", "navigation.tabs.sticky", "navigation.top", "search.highlight", "search.suggest", "toc.follow"], "search": "../assets/javascripts/workers/search.d50fe291.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script> <script src=../assets/javascripts/bundle.13a4f30d.min.js></script> <script src=../assets/_markdown_exec_pyodide.js></script> <script src=../js/timeago.min.js></script> <script src=../js/timeago_mkdocs_material.js></script> <script src=../javascripts/mathjax.js></script> <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script src=https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js></script> </body> </html>